NE-2216: add enhancement for selectable HAProxy version (#1965)
openshift/enhancements: enhancements/ingress/select-haproxy-version.md

### Alternative 2: Automatic Canary Testing

Implement automatic canary testing where the operator gradually rolls out
new HAProxy versions and monitors for issues.

**Why not selected**: While valuable, this is a more complex feature that
could be built on top of this enhancement in the future. The current
proposal provides the building blocks for manual canary testing by allowing
administrators to create separate IngressControllers with different
versions.

### Alternative 3: Complete IngressController Image Selection

Allow selection of entire router images rather than just HAProxy versions.

**Why not selected**: This provides too much flexibility and could lead to
unsupported configurations. The goal is specifically to manage HAProxy
version risk while keeping other router components synchronized with the
OpenShift release.

## Open Questions [optional]

1. Which implementation proposal (1, 2, or 3) should be selected for the
   initial implementation? The choice impacts image size, operational
   complexity, and maintainability. See the Implementation Proposals section
   for detailed trade-offs.

2. How to pin an HAProxy version from OpenShift 4.22 before migrating to 5.0

>    or 5.1?

Author: This is duplicated, same as item 11.

        Operator->>IC: Set degraded condition
        Operator->>Admin: Report error in status
    else Version available
        Operator->>Router: Update deployment with version
        Router->>Router: Restart with new HAProxy binary
        Router->>Operator: Report readiness
        Operator->>IC: Update status with active version
        IC->>Admin: Display updated version in status
    end
```

### API Extensions

This enhancement modifies the existing IngressController CRD
(`operator.openshift.io/v1`) to add a new optional field for specifying
the HAProxy version.

The proposed API fields for the IngressController spec:

```go
// HAProxyVersion specifies which HAProxy version to use for this
// IngressController. Valid values are:
// - "Current" (default): Use the HAProxy version from the current OpenShift
//   release (the most recent version available)
// - "OCP-X.Y": Use the HAProxy version from OpenShift release X.Y
//
// Each OpenShift version has a 1:1 mapping to a specific HAProxy version.
// Exactly 3 versions are supported: the current OpenShift release and the 2
// prior minor releases, provided they are actively supported by Red Hat.
//
// +optional

> // +kubebuilder:default="Current"

Reviewer: I don't think we should be setting defaults explicitly via the API. See https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#defaulting

If unset we can set it via the controller instead here. Alternatively, since this is an optional field, "empty" and "Current" has the same behaviour. Maybe it's not necessary to actually have a "Current" value, and empty value can just be "no opinion"? (i.e. if the user set it at some point and want to unmanage it in the future, they should just remove the spec field entirely)

Author: Makes sense, just updated to drop `Current` as an option, so empty/missing is now the way to represent the default/latest version.

> // +kubebuilder:validation:Pattern=`^(Current|OCP-[0-9]+\.[0-9]+)$`

Reviewer: We should tighten this validation, pending discussions on transition windows and supported versions. We should try to validate as much as we can (preventing downgrades, etc.) here. Trying to think if we can explicitly have set values and transition versions, but maybe that needs to be a VAP instead

Author: @yuqi-zhang my first thought was to use a flexible validation (as the proposal) and a webhook in order to maintain the api stable. At first glance set/enum works for me as well, provided that the API can be updated on every single release to add new options, eventually dropping old ones. Any preferable approach, directions?

cc @gcs278 @Miciah @knobunc

Reviewer: That sounds fine to me, assuming you are ok with updating the supported versions every release on the controller side.

(Also wonder if the `OCP` part is redundant, since we now have OCP in the name, so maybe this can just be 4.22, 5.0, etc.)

Reviewer 2: > We should tighten this validation, pending discussions on transition windows and supported versions. We should try to validate as much as we can (preventing downgrades, etc.) here.

We specifically want to allow downgrades.  If a customer upgrades to OCP 5.0, updates to HAProxy 3.2, and then encounters a critical issue, we want the customer to be able to downgrade back to HAProxy 2.8 while we work on a fix for HAProxy 3.2.  

> Trying to think if we can explicitly have set values and transition versions, but maybe that needs to be a VAP instead

Using a VAP should be sufficient, and much easier than using a webhook or continuously updating the validation in the API definition.  

> @yuqi-zhang my first thought was to use a flexible validation (as the proposal) and a webhook in order to maintain the api stable. At first glance set/enum works for me as well, provided that the API can be updated on every single release to add new options, eventually dropping old ones. Any preferable approach, directions?

> That sounds fine to me, assuming you are ok with updating the supported versions every release on the controller side.

What would be the benefit of using an enum, versus putting a VAP in the openshift/cluster-ingress-operator repository?  

> (Also wonder if the `OCP` part is redundant, since we now have OCP in the name, so maybe this can just be 4.22, 5.0, etc.)

Agreed, if the field name has "OCP" or "Openshift" in it, the field value doesn't need to repeat it.

Author: > Using a VAP should be sufficient

Agreed. I wasn't aware about VAP when I proposed the webhook. Using VAP seems to give both a precise validation and the flexibility to easily update it from the operator repository.

Author: https://github.com/jcmoraisjr/openshift-enhancements/blob/91d29cfae59209b75052dbb786a1cda97048138d/enhancements/ingress/select-haproxy-version.md?plain=1#L611-L619

// +openshift:enable:FeatureGate=SelectableHAProxyVersion

> HAProxyVersion string `json:"haproxyVersion,omitempty"`

Reviewer: Just adding some API notes, since HAProxy has its own version/release, and that's not what we're allowing the user to specify, maybe we should be explicit and name it something like `HAProxyOCPReleaseVersion` (maybe too verbose), or alternatively update the godoc to say OCP release version instead of HAProxy version.

Also this (and the status field) should have min/maxlength markers and godoc.

We should also explicitly have in the godoc what happens if they are not in support (CO's Upgradeable condition is set to False)

Author: Renamed to `HAProxyOCPVersion`, how does it sound? Changed also the godoc to better describe what the field means.

Addressed all the other comments.

Author: We're going to use `HAProxyOCPVersion` then as the field name, and I'm going to update the details about its usage later today.

  promoting to production
- Manage risk more effectively during upgrade cycles
- Maintain stability for critical production applications while exploring
  newer versions

### User Stories

As a cluster administrator, I want to upgrade my OpenShift cluster without
simultaneously upgrading HAProxy on production IngressControllers, so that I
can reduce the risk of application outages during the upgrade process.

As a cluster administrator, I want to create a test IngressController with
a newer HAProxy version, so that I can validate the behavior of the new
version before deploying it to production workloads.

As a cluster administrator, I want to gradually migrate my IngressControllers
from one LTS OpenShift version's HAProxy to the next, so that I can manage
changes incrementally across multiple OpenShift releases.

As a platform operations team, I want to monitor and manage HAProxy versions
across multiple IngressControllers at scale, so that I can ensure
consistency and track version adoption across the cluster.

### Goals

- Enable cluster administrators to select HAProxy versions for individual
  IngressControllers independently of OpenShift cluster version

> - Support exactly 3 distinct HAProxy versions simultaneously: the current

Reviewer 3: Why 3 versions and not 2? Is the rationale to give users a two-release window to migrate?

e.g. HAProxy from 4.22 would be available in 5.0 and 5.1, then dropped in 5.2 - so they have two upgrade cycles to move off it? If so, that should be stated explicitly somewhere in the EP, because the cost of 3 (images, CI, compatibility testing) is higher than 2, and I think we need to capture that somewhere.

Reviewer 3: Actually - should we also specify that we will drop support for EOL HAProxy versions as well? What if the N-2 version is EOL? Will HAProxy 2.8 go EOL before 5.0 is projected to be EOL? https://access.redhat.com/support/policy/updates/openshift#dates might be of interest.

Author: 3 versions is a way to support users that use to skip odd versions and deploy only the LTS ones. @knobunc raised this scenario and can comment about it.

EOL HAProxy is a good point: [2.8 EOL is 2028-Q2](https://www.haproxy.org/) and 5.0 extended is about 2029-Q3 if I'm not mistaken. Wondering if it makes sense to state a specific API value as EOL/Unsupported after some time, so users won't be supported on 5.0 having HAProxy from 4.23 after ~2028-Q2

Reviewer 3: > 3 versions is a way to support users that use to skip odd versions and deploy only the LTS ones. @knobunc raised this scenario and can comment about it.

Got it. Could we add the rationale for 3 versions somewhere in the EP?

> EOL HAProxy is a good point: [2.8 EOL is 2028-Q2](https://www.haproxy.org/) and 5.0 extended is about 2029-Q3 if I'm not mistaken. Wondering if it makes sense to state a specific API value as EOL/Unsupported after some time, so users won't be supported on 5.0 having HAProxy from 4.23 after ~2028-Q2

This is tricky. @Miciah @knobunc we need your input. I think we'd have to deprecate HAProxy 2.8 support in a 5.0 z-stream and set a `Degraded` condition or fire an alert telling users it's no longer receiving CVE fixes and they need to move to 3.2. It's messy, but I don't think extending HAProxy 2.8 support ourselves is realistic. Thoughts?

Reviewer 3: @jcmoraisjr just a nudge on the first part to be added to the EP:

> Got it. Could we add the rationale for 3 versions somewhere in the EP?

Reviewer 2: > Why 3 versions and not 2? Is the rationale to give users a two-release window to migrate?

My original thought was that you would have the previous version (HAProxy 2.8), the current version (HAProxy 3.2), and possibly the next version (HAProxy 3.4?) as a tech-preview option.  This would give customers a way to test out the upcoming version of HAProxy before upgrading to an OCP release where it would be the default version.  

> I think we'd have to deprecate HAProxy 2.8 support in a 5.0 z-stream and set a `Degraded` condition or fire an alert telling users it's no longer receiving CVE fixes and they need to move to 3.2. It's messy, but I don't think extending HAProxy 2.8 support ourselves is realistic. Thoughts?

I don't think it is appropriate to drop support for HAProxy 2.8 in a z-stream.  That feels like a breaking API change and a potential cause of uncertainty and confusion.  We might stop backporting non-critical fixes for the older version, but we should still allow selecting it and continue fixing critical CVEs.

Author: > @jcmoraisjr just a nudge on the first part to be added to the EP:
>
>> Got it. Could we add the rationale for 3 versions somewhere in the EP?

I missed that, just added!

https://github.com/jcmoraisjr/openshift-enhancements/blob/91d29cfae59209b75052dbb786a1cda97048138d/enhancements/ingress/select-haproxy-version.md?plain=1#L630-L640

Author: We've already an agreement Grant described in https://github.com/openshift/enhancements/pull/1965#discussion_r3398963608, so I'm flagging this conversation as resolved.

  OpenShift release and the 2 previous minor releases (e.g., OCP 5.1, 5.0,
  4.22), provided they are actively supported by Red Hat
- Provide a 1:1 mapping from OCP version to HAProxy version (e.g.,
  `haproxyVersion: "OCP-4.22"` always uses the specific HAProxy version that

>   shipped with OCP 4.22)

Reviewer 2: This goal becomes nonsensical once each OCP release ships multiple HAProxy versions.  Rather than "the specific HAProxy version that shipped with OCP" 4._y_, it should be "the specific HAProxy version that _is default_ in OCP" 4._y_.  

I'm still ambivalent about using OCP releases instead of HAProxy releases.  Can you add the reasoning for this goal in the "Alternatives" section, and add a reference to that reasoning here?

Author: Looks good, just updated accordingly.

Also added the alternative to HAProxy related versions and a reason why we chose OCP one instead.

Reviewer 2: To go through an example, with OpenShift 5.0, what are all the allowed values for `haproxyOCPVersion` (as the field is now called), and what versions of HAProxy do those values correspond to?

Reviewer 2: Following up here: https://github.com/openshift/enhancements/pull/1965#discussion_r3353626416

- Allow testing new HAProxy versions on dedicated IngressControllers before
  production deployment
- Maintain compatibility with dynamic HAProxy compilation and required
  dependencies (pcre, openssl, FIPS libraries)
- Default to the HAProxy version from the current OpenShift release when no

>   explicit version is selected

Reviewer 2: We cannot assume that all customers will test before upgrading, or that testing will catch all issues.  Therefore an important story and goal is this:  As a cluster administrator, I want to downgrade an existing IngressController to a previous HAProxy version, so that I can resolve a critical outage or degradation caused by the later HAProxy version.

Author: Good point, added downgrading haproxy version as a new user story.

  itself (router code, templates, etc.)
- Supporting more or fewer than exactly 3 distinct HAProxy versions
  simultaneously (except for OCP 5.0 which supports 2)
- Supporting OCP versions older than current minus 2 (e.g., on OCP 5.1,
  cannot select OCP-4.21 or older)

## Proposal

This enhancement proposes adding a new field to the IngressController API
that allows administrators to specify which HAProxy version to use. The
version is referenced by OpenShift release version (e.g., "OCP-4.21") or
the special value "Current" (default) which always uses the default
HAProxy version on the current OpenShift release.

When an administrator specifies an OpenShift release version, the
IngressController will use the HAProxy version that shipped with that
specific OpenShift release. The ingress-controller-operator will manage the
deployment of the appropriate HAProxy binary and its dependencies (pcre,
openssl, FIPS libraries) to support the selected version.

Exactly 3 distinct HAProxy versions will be supported simultaneously
(current OCP version, previous minor version, and two versions back),
allowing administrators to migrate across OpenShift versions while

> preserving HAProxy versions during the transition period. Each OCP version

Reviewer 2: Specifically, the reason to support exactly _n_-2 is to ease EUS-to-EUS upgrades, correct?

Author: Exactly. The suggestion comes from Ben arguing that several users use to skip non EUS versions.

Reviewer 2: Ah, but OpenShift 5 will not have the same numbering scheme for its Extended Life Cycle (ELC) releases that OpenShift 4 has for its EUS releases; instead, OpenShift 5 will 1 ELC release approximately every _3_ releases (so that it works out to approximately one ELC release every 12 months).  So **if** bridging EUS/ELC releases is required, this might necessitate supporting at least _4_ distinct versions: _n_-3.._n_.

has a 1:1 mapping to a specific HAProxy version. The feature begins with
OCP 5.0.

**Exception**: OCP 5.0 will support only 2 versions (5.0 and 4.22) as it is
the first release where this feature becomes available. Starting with OCP
5.1, exactly 3 versions will be supported (e.g., 5.1, 5.0, and 4.22).

### Workflow Description

**cluster administrator** is a human user responsible for managing
OpenShift cluster infrastructure and upgrades.

#### Selecting a HAProxy Version

1. The cluster administrator reviews the current OpenShift release notes and
   identifies the HAProxy version shipped with the new release.
2. The cluster administrator creates or updates an IngressController resource,
   specifying the desired HAProxy version in the new API field (e.g.,
   `haproxyVersion: "OCP-5.0"` or `haproxyVersion: "Current"`).
3. The ingress-controller-operator validates the requested version is
   available and supported.
4. The operator updates the IngressController deployment to use the
   specified HAProxy version and its matching dependencies.
5. The router pods restart with the selected HAProxy version.
6. The cluster administrator verifies the IngressController status reflects

>    the selected version and routes are functioning correctly.

Reviewer 2: What happens if the cluster-admin repeats this process 3 times?

1. Install OCP 5.0.
2. Specify "OCP-5.0".
3. Upgrade to OCP 5.1.
4. Leave the version at "OCP-5.0".
5. Upgrade to OCP 5.2.
6. Leave the version at "OCP-5.0".
7. Upgrade to OCP 5.3.

Do we block the upgrade (that is, set the ingress clusteroperator's `Upgradeable` status condition to `"False"`) between step 5 and step 7 because OCP 5.3 no longer supports/allows specifying "OCP-5.0"?

Author: Yes, that's the idea. The `Upgradeable` condition should be true only if the current version points to an OCP supported in the version to be upgraded.

// IngressController. Valid values are:
// - "Current" (default): Use the HAProxy version from the current OpenShift
//   release (the most recent version available)
// - "OCP-X.Y": Use the HAProxy version from OpenShift release X.Y
//
// Each OpenShift version has a 1:1 mapping to a specific HAProxy version.
// Exactly 3 versions are supported: the current OpenShift release and the 2
// prior minor releases, provided they are actively supported by Red Hat.
//
// +optional
// +kubebuilder:default="Current"
// +kubebuilder:validation:Pattern=`^(Current|OCP-[0-9]+\.[0-9]+)$`
// +openshift:enable:FeatureGate=SelectableHAProxyVersion
HAProxyVersion string `json:"haproxyVersion,omitempty"`
```

The proposed API fields for the IngressController status:

```go
// EffectiveHAProxyVersion reports the OCP version determining the HAProxy
// version currently in use by this IngressController. This reflects the
// resolved value of the spec.haproxyVersion field.
//
// Examples:
// - "OCP-5.0": Using HAProxy from OpenShift 5.0
// - "OCP-4.22": Using HAProxy from OpenShift 4.22
//
// The actual HAProxy binary version number is available through HAProxy's

> // own metrics.

Reviewer 2: Which specific metrics are you referring to?

Author: This was intended as a label from an existing metric already being used to expose non numeric data. Open question 9 asks about other options, I think a status field is a good addition for a better visibility.

  -f /var/lib/haproxy/conf/haproxy.config
```

Advantages:
- Most explicit control over library loading
- No pollution of environment variables
- Can override RPATH/RUNPATH embedded in binaries
- Compatible with HAProxy's library loading model
- Dynamic linker and libc versions match, avoiding incompatibilities

Disadvantages:
- Platform-specific (loader filename differs across architectures)
- More complex command line
- FIPS compliance verification needed for this approach
- Each version must include its own copy of the dynamic linker
- Running process is the library loader, not haproxy

**Common Advantages for Proposal 1**:
- Single container image to manage
- No changes to pod structure
- Straightforward deployment model
- All versions available immediately

**Common Disadvantages for Proposal 1**:
- Complexity in the library isolation
- Requires special handling if using new features from a newer HAProxy
  version, not available on an older one
- Larger image size (multiple HAProxy binaries and dependencies)
- All versions consume image storage even when unused (no deduplication of
  identical libraries across versions)
- Image rebuilds required to update any HAProxy version

> #### Proposal 2: External HAProxy Images with Sidecar Deployment

Reviewer 3: Sounds like we have landed on this so should update the EP to make this the official proposal and add the others to "alternatives".

Author: Yep, just updated and pushing in a moment. I made it on a separate commit so we preserve the history for now, this can be squashed though before merging.

   All proposals must:
   - Ensure the correct direct dependencies (pcre, openssl, FIPS libraries)
     and indirect dependencies (libc and other system libraries) are available
     for each HAProxy version
   - For Proposal 1: include all dependencies in version-specific directories
     to ensure complete isolation
   - Maintain compatibility matrices for HAProxy versions and their
     dependencies

5. **Version Validation**: Implement validation to ensure:
   - Only supported OpenShift release versions can be specified (current and
     2 previous minor versions)
   - Only "Current" or "OCP-X.Y" format is accepted
   - Requested OCP versions are available in the current cluster
   - Exactly 3 versions are provided (except OCP 5.0 which provides 2)

6. **Upgrade/Downgrade Handling**: Implement logic to handle cluster
   upgrades and downgrades according to the specified version policy:
   - "Current" always uses the latest available version
   - Specific versions are preserved if available in the target release
   - Fall back to default if a version becomes unsupported when
     Downgrading
   - Deny cluster upgrades when any IngressController references a

>      `haproxyVersion` that would become unsupported in the target release,

Reviewer 3: Rather than "deny cluster upgrades" generically, could we use `Upgradeable=False` on the ClusterOperator when an IngressController references the oldest version? Except - that wouldn't happen until 5.1 because we plan to support N-2.

Author: Good point, updated accordingly (pushing in a moment). I'm not sure however how the operator knows about a target version in order to infer if the selected version is supported or not. Any suggestion on how to make these statements more clear?

Reviewer 3: Hmm yea we can work out the implementation details later, but I was thinking something simple like:

If any IngressController references the oldest supported HAProxy version, the operator sets `Upgradeable=False`. Exception: on OCP 5.0, `OCP-4.22` does not block upgrades since it will also be available in 5.1.

It is a bit of implementation detail, so don't worry too much about capturing it perfectly in the EP.

   - On Day 180, if a cluster is running OCP 5.0 with `haproxyVersion:
     "OCP-4.21"`, should the IngressController:
     - Continue running with OCP 4.21's HAProxy?
     - Set a warning/degraded condition?
     - Be forcibly upgraded to a supported version?
   - Does the support matrix update when an OCP version goes EOL, even for
     clusters not yet upgraded to the next version?
   - How do we communicate the support lifecycle to administrators?

7. **Library compatibility across kernel versions**: For Proposal 1
   (particularly 1.a chroot, but applies to all sub-proposals), what are the
   risks of running older dynamic libraries (pcre, openssl, FIPS modules from
   OCP 4.22) on a newer kernel (from OCP 5.1)?
   - Are there known incompatibilities between specific library/kernel
     version combinations?
   - How do we validate compatibility during testing?
   - Should we document supported/tested combinations?

8. **Proposal 3 image sourcing**: For Proposals 2 and 3, when building OCP 5.1
   with three distinct images (containing HAProxy from 5.1, 5.0, and 4.22),
   how are the older image versions obtained?

>    - Can previous OCP releases (5.0 and 4.22) provide their router or HAProxy

Reviewer 3: I think we should reach out to ART about this (if you haven't already) - but I'm pretty certain we shouldn't pull the old images forward - I think we should rebuild them (especially since we are modifying the haproxy sidecar container, aka we'd have to do some backporting). As far as I'm aware, we are considering the previous versions full supported right? I think they need to uphold all of the container updates we have for the current OCP platform version (RHEL 10, OpenSSL, etc...).

Maybe we could use the old RPMs (another ART question) - however, they build with different build containers and there may be compatibility risks. Something we should follow up on.

Reviewer 3: Just saw your [thread](https://redhat-internal.slack.com/archives/CB95J6R4N/p1778760302247079): nice work sorting this out.

So the approach is separate distgit repos per HAProxy version (starting with `haproxy32`), each producing its own RPM, and we build a unique HAProxy sidecar image for each. That makes sense to me so far.

Author: Exactly. I still need to confirm one last item, whether we're going to use one image per haproxy, or single image using `rhel-coreos-extensions`, after that I'll update this one as answered.

     images?
   - What actions are needed to make this approach feasible?

9. **HAProxy version correlation and visibility**: The IngressController
   status shows the OCP version (e.g., `status.effectiveHAProxyVersion:
   "OCP-5.0"`) and HAProxy exposes its own version number in metrics.
   - Do administrators need an explicit mapping table (OCP version → HAProxy
     version number) in documentation?
   - Should the actual HAProxy version number appear in logs, events, or
     additional status fields?

10. **Version support criteria for 5.x releases with 4.x availability**: Once
    OCP 4.23 and newer 4.x versions start releasing (after OCP 5.0 and 5.x are
    available), what should be the criteria for supported HAProxy versions?
    - How does the "current and 2 previous minor versions" rule apply across
      major version boundaries?
    - Should 5.x releases support HAProxy versions from 4.x releases, or only
      from 5.x releases?
    - Example: If OCP 5.2 and OCP 4.25 are both actively supported, should
      OCP 5.2 allow `haproxyVersion: "OCP-4.25"`?
    - Should there be separate support tracks for 4.x and 5.x?

> 11. **HAProxy version pinning during initial feature adoption**: Since OCP 5.0

Reviewer 3: I'd defer to @knobunc and @Miciah here. This one is a tricky one - I lean towards "force an HAProxy upgrade to 3.2" which is the status quo for all OCP minor bumps. But I understand 5.0 is different.

However, you could imagine this situation:
1. user upgrades in a dev cluster - haproxy 3.2 is broken for them
2. user wants to upgrade, but use haproxy 2.8 (4.22)
3. user can't upgrade without first breaking their production cluster, then reverting their ingress controller(s)

I'm not sure. I think it's probably acceptable to say "we're going to bump your HAProxy versions, but you will need to revert them after the upgrade if it doesn't work".

Reviewer 3: @jcmoraisjr I spoke with @Miciah today about this - 3.2 is going to be an opt-in. Upgrades to 5.0 must be low risk - and we cannot auto opt-in users to a new HAProxy version.

You can answer this open question - and answer this in the `Upgrade / Downgrade Strategy section` where you reference this open question.

Author: Just answered it. Question asked about how, so I added a doable approach instead of changing the question. I can change it to skip the details though.

      field that technically works but is not publicly exposed in the API
      schema or documentation?
    - What are the implications for support, security, and FIPS compliance when
      allowing custom images?
    - How does this interact with the supported OCP version mappings?
    - What validations should be applied when a custom image is selected?

## Test Plan

The test plan for this enhancement must include:

**Unit Tests**:
- API validation for the `haproxyVersion` field
- Version selection logic in the operator
- Version compatibility validation
- Upgrade and downgrade scenario handling

**Integration Tests**:
- IngressController creation with different HAProxy versions
- Switching between HAProxy versions on existing IngressControllers
- Behavior during cluster upgrades with various version configurations
- Validation of version limits (maximum 3 distinct versions)

**E2E Tests**:
All E2E tests must include the `[sig-network-edge]` and the
`[OCPFeatureGate:SelectableHAProxyVersion]` labels for the component, and
appropriate test type labels like `[Suite:openshift/conformance/parallel]`,
`[Serial]`, `[Slow]`, or `[Disruptive]` as needed.

Tests must cover:

> - Basic functionality: selecting "Current" and specific OpenShift versions

Reviewer 3: This talks about new `[OCPFeatureGate:SelectableHAProxyVersion]` tests and generically talks about covering basic functionality - but I think we would like as much of our existing E2E test coverage to cover both the current version and previous version(s) right?

Looking at other team that support multiple-versions of a component - the answer seems to be "run multiple jobs". Here's an example from [hypershift](https://github.com/openshift/release/pull/61933). It also looks like they maintain a [matrix](https://github.com/openshift/release/pull/61933/changes#diff-848be3f6e8e9eee2f1898cb9610707fa1fe7f345a75b288bdb607604ccd40360) that auto triggers the appropriate jobs. For example, we could have a installation step that changes the `default` IngressController to a previous version before running the E2E test suite. And that E2E test suite should probably be a subset of the entire origin test suite (i.e. `[sig-network-edge]|[Feature:Router]`).

It'd be best to capture the high level plan here. I think we should also follow up with the TRT team on how they think we should test - this is outside my expertise.

Author: @gcs278 I'm aware you started this via https://github.com/openshift/release/pull/80171, wondering if you can suggest updates in the e2e items that better reflect the approach.

Reviewer 3: sure i can help. Let me get back on suggesting some updates for you.

- Version persistence across cluster upgrades

> - Fallback behavior during downgrades

Reviewer 3: As far as I know, we don't need to support downgrades or test for it.

Author: Ah good to know. We've a few references about downgrades, should them all be removed?

Reviewer 3: I think you can remove them.

- Behavior during cluster upgrades with various version configurations
- Validation of version limits (maximum 3 distinct versions)

**E2E Tests**:
All E2E tests must include the `[sig-network-edge]` and the
`[OCPFeatureGate:SelectableHAProxyVersion]` labels for the component, and
appropriate test type labels like `[Suite:openshift/conformance/parallel]`,
`[Serial]`, `[Slow]`, or `[Disruptive]` as needed.

Tests must cover:
- Basic functionality: using empty value (default) and specific OpenShift versions
- Version persistence across cluster upgrades, using API field on OCP 5.0 or newer
- Version persistence when migrating from 4.22 to 5.0, where API field is not available
- Multiple IngressControllers with different HAProxy versions
- Validation that routes work correctly with different HAProxy versions
- Performance and resource consumption with multiple versions
- HAProxy configuration reload functionality via admin socket API
- Init container startup and static file copying
- Sidecar container communication and shared volume functionality

**Negative Tests**:
- Attempting to use unsupported/unavailable versions
- Invalid version format strings

> 

Reviewer 3: @jcmoraisjr I took a stab a "CI Strategy Section".

To be clear, the sections above cover *what* new tests to write (unit/integration/e2e), but we should also describe *how the existing CI test suite is run against the previous HAProxy version* without writing new tests. It's about taking our existing CI jobs and running them wholesale with the previous HAProxy version.

When drafting this up - I realize it's a bit into the implementation details of setting up the test - so I mean this to be a starting point (hopefully to help whoever implements the testing) rather than a full design. We can make it more generic if it causes a lot of confusion.

Also - I'm assuming 2 HAProxy version (latest and previous) - this is the reason I was interested in not exploding our test matrix to more than 2 because this test matrix gets much bigger.

```suggestion
**CI Strategy**:

All existing CI jobs will implicitly test the default (latest) HAProxy version. To cover the previous HAProxy version, a subset of existing nightly periodic jobs are cloned with a variant suffix (e.g., `-haproxy-prev`) using a CI step registry step (e.g., `ingress-conf-haproxy-version`) that writes an IngressController installer manifest to set the previous HAProxy version at install time. While the feature gate is TechPreviewNoUpgrade, these variants will be based on the `-techpreview` jobs.

These variants will run on a single platform (e.g., AWS) rather than duplicating across every platform, since HAProxy behavior is platform-agnostic. The initial set of previous HAProxy periodic job variants will include:
- E2E conformance (e.g., `e2e-aws-ovn-haproxy-prev`)
- Serial tests (e.g., `e2e-aws-ovn-serial-1of2-haproxy-prev`, `e2e-aws-ovn-serial-2of2-haproxy-prev`)
- Z-stream upgrade (e.g., `e2e-aws-ovn-upgrade-haproxy-prev`)
- Minor version upgrade (e.g., `e2e-aws-ovn-upgrade-major-haproxy-prev`)

Additionally, an optional on-demand presubmit (e.g., `e2e-aws-ovn-haproxy-prev`) will be added to the `openshift/router` and `openshift/cluster-ingress-operator` repositories, allowing developers to trigger it via `/test` when touching HAProxy-related code.

All upgrade variants run continuous ingress disruption monitors throughout the upgrade to detect availability regressions. Note that the upgrade job variants are not needed until OCP 5.0, since the 4.22→5.0 upgrade path defaults to the previous HAProxy (via auto-set `OCP-4.22`) and is already covered by existing major upgrade jobs. If the test matrix grows beyond 2 supported HAProxy versions, parameterized CI workflows will replace manually cloned job variants. This is a starting point — the scope and coverage will be refined as we coordinate with the TRT/TestPlatform team.
```

Author: Perfect, just added it as is.

---
title: select-haproxy-version
authors:
  - "@jcmoraisjr"
reviewers:
  - "@bentito"
  - "@gcs278"
  - "@Miciah"
approvers:
  - "@Miciah"
api-approvers:
  - "@Miciah"
creation-date: 2026-04-01
last-updated: 2026-04-01
tracking-link:
  - https://redhat.atlassian.net/browse/NE-2215
see-also:
replaces:
superseded-by:
---

# Select HAProxy Version

## Summary

This enhancement proposes adding the ability for cluster administrators to
select specific HAProxy versions for IngressControllers, decoupling HAProxy
version upgrades from OpenShift cluster upgrades. This allows administrators

> to test new HAProxy versions independently before deploying them to

Reviewer 4: nit: just to set expectations, as we are only allowing selected versions
```suggestion
to test select new HAProxy versions independently before deploying them to
```

Author: Hmm the new phrasing looked strange to me, maybe you have a new proposal to the whole sentence?

OpenShift environment. They are responsible for exposing applications
outside of the cluster, and even minimal changes during upgrades can lead
to applications misbehaving or experiencing outages. Currently, HAProxy
versions are tightly coupled with OpenShift releases, forcing all
IngressControllers to upgrade HAProxy simultaneously during cluster
upgrades.

By adding the ability to select HAProxy versions, system administrators can:
- Preserve the HAProxy version from a previous OpenShift release during
  cluster upgrades
- Test new HAProxy versions on non-production IngressControllers before
  promoting to production
- Manage risk more effectively during upgrade cycles
- Maintain stability for critical production applications while exploring
  newer versions

### User Stories

As a cluster administrator, I want to upgrade my OpenShift cluster without
simultaneously upgrading HAProxy on production IngressControllers, so that I
can reduce the risk of application outages during the upgrade process.

As a cluster administrator, I want to create a test IngressController with
a newer HAProxy version, so that I can validate the behavior of the new

> version before deploying it to production workloads.

Reviewer 4: nits, but provide clarification and set expectations (we should already have validated it, they are welcome to evaluate it for their purposes).
```suggestion
As a cluster administrator, I want to create a sharded test IngressController with
a newer HAProxy version, so that I can evaluate the behavior of the new
version before deploying it to production workloads, or evaluate the behavior in canary test fashion on a small subset of traffic.
```

Reviewer 2: It isn't necessary to configure sharding to use this version selection API for any of these purposes.

As a cluster administrator, I want to gradually migrate my IngressControllers
from one LTS OpenShift version's HAProxy to the next, so that I can manage
changes incrementally across multiple OpenShift releases.

As a cluster administrator, I want to downgrade an existing IngressController
to a previous HAProxy version, so that I can resolve a critical outage or
degradation caused by a later HAProxy version.

As a platform operations team, I want to monitor and manage HAProxy versions
across multiple IngressControllers at scale, so that I can ensure
consistency and track version adoption across the cluster.

### Goals

- Enable cluster administrators to select HAProxy versions for individual
  IngressControllers independently of OpenShift cluster version
- Support exactly 3 distinct HAProxy versions simultaneously: the current
  OpenShift release and the 2 previous minor releases (e.g., OCP 5.1, 5.0,

>   4.22), provided they are actively supported by Red Hat

Reviewer 4: I saw the comments above about the number of versions supported simultaneously, but I also want reiterate @gcs278's point that we don't always upgrade HAProxy versions each OCP release.  

What if we instead choose to support 2 versions of HAProxy per OCP release, but we select these versions and they can be replaced over time.  (This is how I thought we were doing this originally - offer the latest version and the last known working version on each OCP release).

For example, what if we chose an HAProxy release 4.1.2 that we later had to make obsolete due to a major vulnerability or other blocker?  We don't want to lose our ability to update an HAProxy release in a z-stream if needed (and in this case, drop 4.1.2 from the selection list to replace it with its fixed version).  This would be better for users too, because we drop broken versions from the upgrade graph automatically.

Author: I'm not sure if I got all the details of your concern, so let me comment based on my understanding on how we're dealing with them.

> What if we instead choose to support 2 versions of HAProxy per OCP release ...

We are using 3 so we can assign one per OCP release, and users have the ability to skip non LTS ones and also be able to upgrade OCP without upgrading HAProxy.

> ... but we select these versions and they can be replaced over time

We can replace HAProxy over time, both the default one and the V-1 and V-2 ones, since they come from a live branch from the `haproxyNN` repos. We update there, they are replaced on all OCP releases. Of course that, the older the release, the higher the bar to allow and validate bumps and patches.

>  We don't want to lose our ability to update an HAProxy release in a z-stream if needed

We won't. Our haproxy image builds are tied with dist-git repos and live branches, so any updates on the branches should end up updating the images and the haproxy binary added to the OCP bundle.

Please let me know if I'm missing something.

- Provide a 1:1 mapping from OCP version to the default HAProxy version for
  that release (e.g., `haproxyVersion: "OCP-4.22"` always uses the default
  HAProxy version from OCP 4.22), providing an intuitive correlation between
  OCP versions and HAProxy versions (see Alternative API Approach #2 for a
  rejected approach using abstract version identifiers)
- Allow testing new HAProxy versions on dedicated IngressControllers before
  production deployment
- Maintain compatibility with dynamic HAProxy compilation and required
  dependencies (pcre, openssl, FIPS libraries)
- Default to the HAProxy version from the current OpenShift release when no
  explicit version is selected

### Non-Goals

> - Backporting HAProxy versions from newer OpenShift releases to older

Reviewer 4: Excellent!  This is a 5.0 feature, and a great motivation for users to keep up to date.

  dependencies (pcre, openssl, FIPS libraries)
- Default to the HAProxy version from the current OpenShift release when no
  explicit version is selected

### Non-Goals

- Backporting HAProxy versions from newer OpenShift releases to older
  clusters
- Providing formal support for custom HAProxy versions or images not shipped
  with supported OpenShift releases (though the mechanism may technically
  allow custom versions in an unsupported configuration - see Open Question #12)
- Allowing selection of specific HAProxy version numbers (e.g., "2.8.5") in
  the standard supported configuration - only OCP version references are
  supported (e.g., "OCP-4.22")
- Providing version selection for other ingress components beyond HAProxy
  itself (router code, templates, etc.)
- Supporting more or fewer than exactly 3 distinct HAProxy versions
  simultaneously (except for OCP 5.0 which supports 2)
- Supporting OCP versions older than current minus 2 (e.g., on OCP 5.1,
  cannot select OCP-4.21 or older)

## Proposal

This enhancement proposes adding a new field to the IngressController API
that allows administrators to specify which HAProxy version to use. The

> version is referenced by OpenShift release version (e.g., "OCP-4.21") or

Reviewer 3: We often ship OpenShift releases with the same HAproxy version. For example, 4.17-4.20 is all 2.8.10. That's 4 releases with the same version.

1. If multiple OCP releases ship the same HAProxy version, users see different API values (`OCP-5.1`, `OCP-5.2`) that resolve to the same binary. Is that confusing?
2. Does this feature motivate us to bump HAProxy more frequently to make version selection actually useful? If we go several releases without a new HAProxy version, the feature exists but provides no real choice.

Author: I'm not sure having e.g. `OCP-5.1` and `OCP-5.2` pointing to the exact same haproxy binary/image a problem, but instead an opportunity for us to change this reality.

~This scenario brings however another concern to me. We're going to add haproxy repos per haproxy release branch (2.8, 3.2), but what if OCP5.1 and 5.2 both use haproxy32 and the version was either bumped or patched? Maybe we should create parallel repos per ocp release instead?~

* ~`haproxy50` repo, `ocp50` branch: bits of `OCP-5.0` selected version, released on `OCP-5.0` bundle~
* ~`haproxy50` repo, `ocp51` branch: bits of `OCP-5.0` selected version, released on `OCP-5.1` bundle~
* ~`haproxy51` repo, `ocp51` branch: bits of `OCP-5.1` selected version, released on `OCP-5.1` bundle~

~Repo name has the bits of the version user selects via the new API, branch name is the OCP product where those bits should be shipped to (compatible with other repos).~

Reviewer 2: > We're going to add haproxy repos per haproxy release branch (2.8, 3.2), but what if OCP5.1 and 5.2 both use haproxy32 and the version was either bumped or patched? Maybe we should create parallel repos per ocp release instead?

The way dist-git works is that there is one repository with branches for each combination of OCP release and RHEL version (for example, we have both "rhaos-4.17-rhel-8" and "rhaos-4.17-rhel-9").  It makes sense to me to extend this to a branch for each combination of OCP release, RHEL version, and HAProxy version.

Author: We agreed on this matter to:

* Use distinct dist-git repos to handle parallelism, since it's harder for the ART team to handle it changing the branch name pattern;
* Use the `haprox32`, `haproxy34`, ..., which makes it simpler and easier to correlate with haproxy minors. This pattern will make us to have haproxy patched on a single repo and branch, but bumped on up to three distinct OCP releases, since newer OCP releases also point to haproxy versions on older releases.

Reviewer 4: Did you drop @Miciah 's suggestion for a 
> combination of OCP release, RHEL version, and HAProxy version?

Otoh, I still think the name shouldn't be tied to OCP version since we don't always need to update haproxy on each OCP release, while at the same time I propose that we don't want to lose the ability to publish more than one haproxy release per OCP release if we need to drop a bad haproxy release.  See https://github.com/openshift/enhancements/pull/1965/changes#r3344401390.

Author: > Did you drop Miciah's suggestion

Not exactly, but my last message maybe wasn't that clear. ART asked us to handle parallelization via distinct repos instead of changing the branch naming. This would diverge on how other repository's branches are named, making it harder to work on our haproxy particularities.

> I still think the name shouldn't be tied to OCP version ...

Being tied to OCP is a choice we made and I added a rationale here:

https://github.com/jcmoraisjr/openshift-enhancements/blob/455df95271b3bbfe0e897efa58b7ec8254276b16/enhancements/ingress/select-haproxy-version.md?plain=1#L705-L717

```
#### Alternative 2: Limited Number of HAProxy Versions

Use a limited number of HAProxy version identifiers (e.g., "2.8.10", "2.8.18",
"3.2.19") instead of OCP version references. Each identifier would map to a
specific HAProxy version available in the current OCP release.

**Why not selected**: This approach partially solves the difficulties of
Alternative 1 (arbitrary HAProxy version numbers) by limiting the support
matrix to a fixed number of versions. However, it does not provide an easy
or intuitive correlation to the OCP version currently running. Administrators
would need to consult documentation or status fields to understand which
actual HAProxy version or OCP release each identifier represents, making it
less transparent than the OCP-based referencing approach.
```

This is not a final decision though, but I'd add @knobunc in the discussion since the idea came from him.

> ... since we don't always need to update haproxy on each OCP release

This shouldn't be a reason I think, and also, this shouldn't happen because HAProxy always have at least patches under a minor release, and since we're adopting an opt-in migration, we don't have a reason anymore to not bump HAProxy, either minor or patch, between OCP releases.

Reviewer 3: I agree x.y.z is not maintainable, but did we ever consider selecting HAProxy version based on x.y (major.minor)? e.g. `2.8` or `3.2` and the z-streams are floating based on the OCP version (similar to how they are today). This would be clearer to the user and avoids the confusing situation where multiple OCP versions map to the same HAProxy version.

From a user perspective, it's more ~~intuitive~~ clear/explicit to say "I want HAProxy 2.8" than "I want the HAProxy version that shipped with OCP 4.22." It's kinda of similar to how dual-stream RHCOS works ([OCPSTRAT-1150](https://redhat.atlassian.net/browse/OCPSTRAT-1150)): clusters select RHEL 9 or RHEL 10 by OS major version, not by which OCP release introduced it.

The one drawback of `x.y` is that it wouldn't enable users to pin to a specific z-stream. However, if the previous HAProxy versions are fully supported (I think they are), then they need CVE patches, which often come in the form of z-stream bumps. That may be a moot point - z-stream pinning for haproxy isn't really realistic or a good idea.

Maybe I'm thinking too much like an engineer here, but why not base the selector on HAProxy's major.minor (e.g. `2.8`, `3.2`) instead of OCP versions? Users who don't want to pin could just leave the field empty to get the latest (or maybe `latest` as a tag). It'd nicely align to our RPM repo structure too.

Reviewer 2: > > ... since we don't always need to update haproxy on each OCP release
> 
> This shouldn't be a reason I think, and also, this shouldn't happen because HAProxy always have at least patches under a minor release, and since we're adopting an opt-in migration, we don't have a reason anymore to not bump HAProxy, either minor or patch, between OCP releases.

This statement worries me a little.  The primary reason we don't bump HAProxy versions every OpenShift minor release that we have limited engineering capacity to evaluate and document upstream changes, update the RPM, and run regression tests.  Even with this version-selection API, we should still perform those steps diligently; otherwise, we would be shifting the burden of testing and finding issues onto the end-user.  Moreover, we would be setting the stage for a bad situation if, say, HAProxy 3._y_ introduced a performance regression and we didn't manage to fix it before it were time to drop support for HAProxy 3.(_y_-1).  

> I agree x.y.z is not maintainable, but did we ever consider selecting HAProxy version based on x.y (major.minor)? e.g. `2.8` or `3.2` and the z-streams are floating based on the OCP version (similar to how they are today). This would be clearer to the user and avoids the confusing situation where multiple OCP versions map to the same HAProxy version.

I agree, including the _z_ stream would be untenable.  This is what I originally wrote in [NE-2215](https://redhat.atlassian.net/browse/NE-2215):

> Should we allow specifying a specific HAProxy major version ("HAProxy2", "HAProxy3"), minor version ("HAProxy28", "HAProxy32"), stream (for example, "previous", "stable", "experimental"), or something else?

> From a user perspective, it's more ~intuitive~ clear/explicit to say "I want HAProxy 2.8" than "I want the HAProxy version that shipped with OCP 4.22." It's kinda of similar to how dual-stream RHCOS works ([OCPSTRAT-1150](https://redhat.atlassian.net/browse/OCPSTRAT-1150)): clusters select RHEL 9 or RHEL 10 by OS major version, not by which OCP release introduced it.

Agreed.  It's also how Sail Operator's `Istio.spec.version` field works (though that *does* allow specifying patch versions, which I agree is untenable for us).  

> Maybe I'm thinking too much like an engineer here, but why not base the selector on HAProxy's major.minor (e.g. `2.8`, `3.2`) instead of OCP versions? Users who don't want to pin could just leave the field empty to get the latest (or maybe `latest` as a tag). It'd nicely align to our RPM repo structure too.

+1, but I would like to better understand the reasoning for using OpenShift release numbers.  Maybe the end-user or support engineer really is more likely to think, "I want the proxy that came with OCP 4.22", rather than, "I want HAProxy 2.8"?

Reviewer 3: > +1, but I would like to better understand the reasoning for using OpenShift release numbers.

I think the intent is "we support the previous EUS release's HAProxy version." But HAProxy moves much slower than OCP — it's likely that multiple consecutive OCP releases ship the same HAProxy minor. Even though we'd maintain separate RPM builds per OCP version, we'd (probably) be bumping them to the same z-stream for CVE fixes and ART will bump the underlying container image — so "pin to the previous EUS" and "use current" give you the exact same HAProxy binary and image. The version selection API provides no meaningful choice in those cycles, while still adding user-facing API values that resolve to the same binary and unnecessary `Upgradeable=False` friction when those redundant values fall off the N-2 window.

I think the commitment should be framed around HAProxy minor versions, not OCP versions it'd simplify and make the implementation more clear:

_**When we introduce a new HAProxy minor version, the previous minor remains available for at least one EUS cycle to give customers time to validate and migrate.**_

This also different from "we support N-1 HAProxy minor versions" — because if we go several OCP releases without bumping the HAProxy minor, that N-1 HAProxy version stays in the support window indefinitely. It could still be there long after it's gone EOL upstream.

In cycles where we don't bump the HAProxy minor, there's effectively one version to select. The user sees that HAProxy hasn't changed (aside from z-stream patches and container base updates), and the version selection API stays quiet — no `Upgradeable=False`, no migration needed, no action required.

The drawback of HAProxy-version-based selection is that it doesn't encapsulate the full dependency stack. If a customer wants to pin not just HAProxy 3.2 but the specific OpenSSL version, RHEL base, glibc, PCRE compilation, or FIPS module that shipped with a particular OCP release, a HAProxy version label doesn't express that. The OCP-version approach implicitly encapsulates all of those — but in practice we'd be rebuilding old HAProxy RPMs on the current base image anyway (right?), so even that doesn't truly preserve the original stack. And most of these dependency decisions aren't optional for us — they're mandated by ART (current RHEL base, current OpenSSL, etc.). Worth considering, though I'm not sure we can actually deliver a true OCP-version pin on the full dependency stack.

Author: > I would like to better understand the reasoning for using OpenShift release numbers

My rationale on it, despite on how the user should better correlate the proxy version, is that OCP based versions help to achieve the pin of the haproxy patch (x.y.z) and manual patches. As an example, the 2.8.10 to 2.8.18 bump we did from 4.21 to 4.22 should be `4.21` versus `4.22` on OCP based naming, and `2.8` only on HAProxy based one.

This should happen frequently on new OCP minors because we release more often than HAProxy LTS (once a year) and even HAProxy minors (twice a year).

I'm not sure how this could be achieved using haproxy based versions, except if using the patch release (x.y.z) in the version selection.

> it doesn't encapsulate the full dependency stack

Exactly. Our version selection works more or less like this: "User moves from `4.22.5` to `5.0.1` and has an issue in the ingress, and revert to `4.22` / `haproxy28`. This does not revert back to `4.22.5`, but instead to some newer `4.22.z`."

Author: It seems that the core question on this is: do we need to pin haproxy on z-streams while bumping it on newer OCP versions that share the same haproxy repo? If so, using haproxy-version model will not provide information enough for users that prefer stability --- `HAProxy-3.2` version would not represent the same pinned version between those OCP releases, while `OCP-5.0` represents the pinned version on that z-stream.

Otoh, haproxy-verson model should be fine if pinned z-streams means haproxy version frozen on all OCP releases sharing the same haproxy repo, or if bumping haproxy version on a newer OCP release means bumping it as well on all z-streams sharing the same haproxy repo.

Reviewer 3: Summarizing our discussion today:

After discussing with @Miciah  and @jcmoraisjr, we're proposing to simplify the version support model to an MVP: each release ships at most 2 HAProxy versions, with 3 user behaviors.

The current EP's N-2 model would require up to 4 HAProxy versions per release (since ELC is every 3 releases in 5.x), creating a large test and support matrix plus compat distgit repos for each pinnable version. With this proposal, a release only ships the current HAProxy version and the previous ELC release's HAProxy version. Non-ELC releases still bump z-streams normally but don't add new pin targets.

From those 2 versions, users get 3 behaviors:
- **`""` or `latest` (unpinned)** — always on the latest HAProxy, automatically upgrades across OCP releases
- **Pinned to current ELC (e.g. `OCP-5.4`)** — stays on the current HAProxy version, won't auto-upgrade when moving to the next release
- **Pinned to previous ELC (e.g. `OCP-5.1`)** — stays on the previous ELC's HAProxy, for users who just upgraded and aren't ready to move forward yet

This also reduces the amount of distgit repos needed — one compat (frozen) repo per ELC instead of per OCP release. If we later need intermediate pins, we can add them without breaking anything.

Author: Just updated to address the "max of 2 haproxy images per release" approach. There's a separate commit having only this change.

      - Defaults to "Current" and also pre-populate the API to "OCP-4.22"?
      - Or should it default to OCP-4.22 exceptionally on 5.0, changing the
        default to "Current" only on newer versions?
    - Is there a migration window or grace period where administrators can
      change the API field before the routers are updated?
    - How do we document this bootstrapping scenario for the first adopters?

12. **Custom HAProxy image configuration**: Should administrators be allowed to
    configure their own HAProxy image in an unsupported fashion, instead of
    always relying on the OCP-related ones?
    - Could this be done via a documented but unsupported API field that is
      officially exposed but carries no support guarantees?
    - Alternatively, should this be implemented as a hidden and undocumented
      field that technically works but is not publicly exposed in the API
      schema or documentation?
    - What are the implications for support, security, and FIPS compliance when
      allowing custom images?
    - How does this interact with the supported OCP version mappings?
    - What validations should be applied when a custom image is selected?

> 

Reviewer 3: I have a concern - could be another open question:

Is the sidecar update too risky to be the default upgrade path in 5.0? Should we de-risk 5.0 further by keeping the monolithic router+HAProxy deployment as the default path, and only use the sidecar topology when a user explicitly selects a new HAProxy version?

Curious to hear what @Miciah  and @knobunc have to say. I agree maintaining both paths isn't ideal, but given the emphasis on not breaking 5.0 upgrades, the deployment refactoring concerns me.

Reviewer 2: When you say "maintaining both paths", you refer to significantly different paths in both openshift/router#772 and openshift/cluster-ingress-operator#1439, right?  I would rather get the necessary refactorings in place piecemeal, *not* behind a featuregate, to maximize CI soak time and minimize the complexity of supporting multiple paths across multiple components.

Reviewer 2: Keep in mind, we need to maintain these multiple paths as long as the featuregate exists (even after promotion to the "Default" feature set).

Reviewer 3: > When you say "maintaining both paths", you refer to significantly different paths in both openshift/router#772 and openshift/cluster-ingress-operator#1439, right?

To be clear, I'm not hung up on the feature gate — it doesn't de-risk the release since it's either on or off at GA. What I was suggesting is that we could have two deployment architectures without a feature gate: monolithic for HAProxy 2.8 (the `OCP-4.22` selection) and sidecar for HAProxy 3.2+ (the `Current` or newer selections). That way, the standard 4.22→5.0 upgrade path stays monolithic — the sidecar only kicks in when someone opts into a different HAProxy version, de-risking the default upgrade.

It's more effort to maintain two code paths, but I don't think it'd be that bad — the router [PR](https://github.com/openshift/router/pull/772) is already structured with both paths (`reloadRouterEmbedded` vs `reloadRouterExternal`). That said, I don't feel strongly. If we're confident in a full migration to sidecar for all HAProxy versions in 5.0, I'm okay with that too.

Reviewer 3: To be clear, I don't feel strongly about this. If @jcmoraisjr and @Miciah are both comfortable with the sidecar being the only path, I'm fine with that.

dedicated container image, providing clean separation of concerns and
independent versioning. See the Implementation section below for detailed
information.

### Workflow Description

**cluster administrator** is a human user responsible for managing
OpenShift cluster infrastructure and upgrades.

#### Selecting a HAProxy Version

1. The cluster administrator reviews the current OpenShift release notes and
   identifies the HAProxy version shipped with the new release.
2. The cluster administrator creates or updates an IngressController resource,
   specifying the desired HAProxy version in the new API field (e.g.,
   `haproxyOCPVersion: "OCP-5.0"`, or leaving it empty for the default).
3. The ingress-controller-operator validates the requested version is
   available and supported.
4. The operator updates the IngressController deployment to use the
   specified HAProxy version and its matching dependencies.
5. The router pods restart with the selected HAProxy version.
6. The cluster administrator verifies the IngressController status reflects
   the selected version and routes are functioning correctly.

#### Testing a New HAProxy Version Before Production

> 1. The cluster administrator creates a new IngressController with

Reviewer 4: ```suggestion
1. The cluster administrator creates a new sharded IngressController with
```

   `haproxyOCPVersion` unset (empty) to test the latest HAProxy version.
2. The cluster administrator configures test routes to use the new
   IngressController via domain or namespace selectors.
3. The cluster administrator runs tests against the test IngressController
   to validate HAProxy behavior.
4. Once validated, the cluster administrator updates production
   IngressControllers to use an empty `haproxyOCPVersion` (for the latest) or
   a specific OpenShift version.

#### Upgrading OpenShift with HAProxy Version Control

1. The cluster administrator initiates an OpenShift cluster upgrade from
   version 5.0 to 5.1.

> 2. For IngressControllers with `haproxyOCPVersion` unset (empty), the operator

Reviewer 4: Is there a more user-friendly way to do this?  These procedures mandate that the cluster admin checks/sets the haproxyOCPVersion both before and after upgrade if they want to do a cluster upgrade with just one IC taking the new version.

Maybe another api field to update the haproxy version on cluster upgrade if set?  "AutoUpgrade" (default false) for example?

Author: I'm not sure that an `AutoUpgrade` option would help, this is the behavior of leaving the field empty and would cause some confusion. But maybe we can add instead a default value for all IngressController on a higher level/api, how about it?

Need however some input from folks more skilled on the APIs to verify if this is a doable approach.

OpenShift cluster infrastructure and upgrades.

#### Selecting a HAProxy Version

1. The cluster administrator reviews the current OpenShift release notes and
   identifies the HAProxy version shipped with the new release.
2. The cluster administrator creates or updates an IngressController resource,
   specifying the desired HAProxy version in the new API field (e.g.,
   `haproxyOCPVersion: "OCP-5.0"`, or leaving it empty for the default).
3. The ingress-controller-operator validates the requested version is
   available and supported.
4. The operator updates the IngressController deployment to use the
   specified HAProxy version and its matching dependencies.
5. The router pods restart with the selected HAProxy version.
6. The cluster administrator verifies the IngressController status reflects
   the selected version and routes are functioning correctly.

#### Testing a New HAProxy Version Before Production

1. The cluster administrator creates a new sharded IngressController with
   `haproxyOCPVersion` unset (empty) to test the latest HAProxy version.
2. The cluster administrator configures test routes to use the new

>    IngressController via domain or namespace selectors.

Reviewer 2: ```suggestion
   IngressController via route or namespace label selectors.
```

   preserves the HAProxy version from OpenShift 5.0.
4. The cluster administrator validates production applications on
   IngressControllers running HAProxy from OpenShift 5.0.
5. The cluster administrator gradually updates production IngressControllers
   to use newer HAProxy versions after validation.

```mermaid
sequenceDiagram
    participant Admin as Cluster Administrator
    participant IC as IngressController Resource
    participant Operator as Ingress Operator
    participant Router as Router Pods

    Admin->>IC: Update haproxyOCPVersion field
    IC->>Operator: Watch event for spec change
    Operator->>Operator: Validate version availability
    alt Version not available
        Operator->>IC: Set degraded condition
        Operator->>Admin: Report error in status
    else Version available
        Operator->>Router: Update deployment with version
        Router->>Router: Restart with new HAProxy binary
        Router->>Operator: Report readiness
        Operator->>IC: Update status with active version
        IC->>Admin: Display updated version in status
    end
```

### API Extensions

This enhancement modifies the existing IngressController CRD

> (`operator.openshift.io/v1`) to add a new optional field for specifying

Reviewer 2: ```suggestion
(`ingresscontrollers.operator.openshift.io/v1`) to add a new optional field for specifying
```

the HAProxy version.

The proposed API fields for the IngressController spec:

```go
// HAProxyOCPVersion specifies the OCP release whose default HAProxy version
// should be applied to this IngressController.
//
// For example, selecting "OCP-4.22" will configure the IngressController to use 
// the default HAProxy version shipped with OCP 4.22.
//
// Valid values are:
// - Empty string (default): Use the default HAProxy version from the current

> //   OpenShift release (the most recent version available)

Reviewer 2: "(the most recent version available)" is incorrect for OpenShift 5.0.  

Edit: Never mind, I see now that you plan to have some migration logic to set a value automatically during the 4.22→5.0 upgrade (see https://github.com/openshift/enhancements/pull/1965#discussion_r3353626416).

the primary use case. Administrators can manage HAProxy versions
independently across multiple IngressControllers.

#### Single-node Deployments or MicroShift

For Single-Node OpenShift (SNO) deployments, this enhancement provides the
same benefits as standalone clusters. Resource consumption is minimal as
only the selected HAProxy binary and its dependencies are loaded.

For MicroShift, this enhancement may not be directly applicable as
MicroShift has a different ingress architecture. If MicroShift adopts
IngressController resources in the future, this enhancement could be
extended to support it.

#### OpenShift Kubernetes Engine

This enhancement works with OpenShift Kubernetes Engine (OKE) as it relies
on standard IngressController resources which are available in OKE.

### Implementation Details

#### Chosen Implementation: External HAProxy Images with Sidecar Deployment

This implementation deploys HAProxy as a separate sidecar container alongside the
main router container. Each supported HAProxy version is packaged in its own
dedicated container image. One HAProxy sidecar image is built per OCP version,

> with a 1:1 mapping between OCP version and HAProxy version.

Reviewer 2: The wording here is confusing.  We will have one container image for each supported HAProxy minor release for each OpenShift release—right?

**Pod Structure**:
```yaml
spec:
  initContainers:
  - name: init-router-config
    image: registry.redhat.io/openshift4/ose-haproxy-router:v5.0
    # Run script to copy static files (error pages) and templates to shared volume
    command: ["/usr/local/bin/init-haproxy-files.sh"]
    volumeMounts:
    - name: haproxy-shared
      mountPath: /mnt/shared
  containers:
  - name: router
    image: registry.redhat.io/openshift4/ose-haproxy-router:v5.0
    # Router logic, template rendering, route watching
    volumeMounts:
    - name: haproxy-shared
      mountPath: /var/lib/haproxy
  - name: haproxy
    image: registry.redhat.io/openshift4/ose-haproxy:4.22
    # HAProxy binary with its dependencies
    volumeMounts:
    - name: haproxy-shared
      mountPath: /var/lib/haproxy
  volumes:
  - name: haproxy-shared
    emptyDir: {}
```

The init container runs from the router image and executes a shell script to
copy static files (error pages, scripts) and HAProxy configuration templates
from the router image filesystem to the shared emptyDir volume. The router

> container generates HAProxy configuration and writes it to the shared volume.

Reviewer 2: You could put this setup in a script in the proxy image instead of in an init container, but I don't know if that would ultimately be an improvement.

Author: It'd be an improvement by simplifying the deployment topology. But otoh:

* we'd be just changing where the complexity is;
* we cannot copy since the router image is not in place when the haproxy one starts - we are using native sidecars;
* the init container mounts the scripts/confs on a distinct dir, without hiding/overriding the original content, giving the router and haproxy the ability to mount it exactly in the same place they already expect on the embedded/monolith version.

The router communicates with HAProxy through the HAProxy admin socket (also
on the shared volume) to trigger configuration reloads and manage the HAProxy
process.

**Advantages**:
- Clean separation of concerns (router logic vs HAProxy runtime)
- Smaller individual images
- Independent versioning and updates
- No library isolation complexity
- Only selected version image is pulled

> - HAProxy image can be updated independently

Reviewer 2: Is "HAProxy image can be updated independently" redundant with "Independent versioning and updates"?

Author: I don't remember anymore what I meant with the later, just dropped it.

**Disadvantages**:
- Requires pod structure changes (init container + sidecar)

> - Minimal additional container overhead

Reviewer 2: How is this a disadvantage?  ...Ah, are you talking about the marginal added cost of running two containers inside a pod instead of one?  The wording is confusing, and the cost should be negligible.

Author: Just rephrase instead of drop, since it is a disadvantage although a negligible one - wdyt?

> - Minimal overhead in the additional sidecar container

- Running process is the library loader, not haproxy

**Common Advantages for Proposal 1**:
- Single container image to manage
- No changes to pod structure
- Straightforward deployment model
- All versions available immediately

**Common Disadvantages for Proposal 1**:
- Complexity in the library isolation
- Requires special handling if using new features from a newer HAProxy
  version, not available on an older one
- Larger image size (multiple HAProxy binaries and dependencies)
- All versions consume image storage even when unused (no deduplication of
  identical libraries across versions)
- Image rebuilds required to update any HAProxy version

#### Alternative 2: Distinct Router Images per HAProxy Version

This proposal creates completely separate router images for each supported
HAProxy version. Each image contains the router code and a single embedded
HAProxy version with its dependencies. All images are built together from the
same source during each OCP release, with only the embedded HAProxy binary
differing. Images are tagged by the target OCP version.

The router and HAProxy evolve together within each image as a complete unit.

> The ingress-controller-operator selects the appropriate image based on the

Reviewer 2: ```suggestion
The cluster-ingress-operator selects the appropriate image based on the
```
I see "ingress-controller-operator" in a few places; please do a search-and-replace of all instances.

   feature gate in
   https://github.com/openshift/api/blob/master/features/features.go with
   the `TechPreviewNoUpgrade` feature set. The feature gate must
   specify the Jira component, contact person, and link to this enhancement.

3. **Operator Logic**: Update the ingress-controller-operator to:
   - Read and validate the `haproxyOCPVersion` field
   - Map the OCP version to the appropriate HAProxy sidecar container image
     reference
   - Update the router deployment to include the HAProxy sidecar container
     with the selected image
   - Configure the init container to copy static files and templates to the
     shared volume
   - Report the effective HAProxy version in a new IngressController status
     field (e.g., `status.effectiveHAProxyOCPVersion: "OCP-4.22"`) showing the
     OCP version that determines the HAProxy version in use
   - HAProxy's own version number is available through HAProxy's built-in
     metrics

4. **HAProxy Image Management**: Using the sidecar deployment model:
   - Build separate HAProxy container images for each supported OCP version
   - Each HAProxy image includes the HAProxy binary and all its dependencies

>      (pcre, openssl, FIPS libraries, libc, and other system libraries)

Reviewer 2: A bit part of this enhancement (and a lot of the discussion going into it) concerns building the RPMs that go into these images.  Will you be adding the relevant details in this EP?

Author: I'm not sure how much details we do expect in the EP itself, but in fact I think I can go a bit deeper into the details.

Author: Added as a new item here:

https://github.com/jcmoraisjr/openshift-enhancements/blob/91d29cfae59209b75052dbb786a1cda97048138d/enhancements/ingress/select-haproxy-version.md?plain=1#L601-L603

   - The operator maintains a mapping from OCP version (e.g., "OCP-4.22") to
     the corresponding HAProxy sidecar image reference
   - Maintain compatibility matrices for HAProxy versions and their
     dependencies

5. **Version Validation**: Implement validation to ensure:
   - Only supported OpenShift release versions can be specified (current and
     2 previous minor versions)
   - Only empty string or "OCP-X.Y" format is accepted
   - Requested OCP versions are available in the current cluster
   - Exactly 3 versions are provided (except OCP 5.0 which provides 2)

6. **Upgrade Handling**: Implement logic to handle cluster upgrades according
   to the specified version policy:

>    - Empty `haproxyOCPVersion` always uses the latest available version

Reviewer 2: ...except on at least OCP 5.0.  

Edit: Never mind, I see now that you plan to set a value automatically during the 4.22→5.0 upgrade so it will not be empty (see https://github.com/openshift/enhancements/pull/1965#discussion_r3353626416).

   - Specific versions are preserved if available in the target release
   - Set the ingress ClusterOperator resource's `Upgradeable` condition to
     `False` when any IngressController references a `haproxyOCPVersion` that
     would become unsupported in the target release, preventing cluster
     upgrades until administrators update to empty or a supported version.

### Risks and Mitigations

**Risk 1**: Supporting multiple HAProxy versions increases operational complexity
with the sidecar deployment model.

**Mitigation**: Limit support to 3 distinct versions. The sidecar approach
minimizes individual image sizes and only pulls the selected version. Monitor

> image size and establish clear deprecation policies. The init container and

Reviewer 2: "Monitor image size" seems unnecessary.  However, "establish clear deprecation policies" is important.

sidecar pattern is well-established in Kubernetes, reducing operational risk.

**Risk 2**: Administrators may select outdated HAProxy versions with known
security vulnerabilities.

**Mitigation**: Clearly document supported versions and deprecation
timelines. Provide warnings in the IngressController status when using
older versions. Consider implementing alerts when versions reach

> end-of-support.

Reviewer 2: No, we should fix those security vulnerabilities.

Reviewer 3: Yea one thing I was trying to clarify — in the OCP-version model, are we actually pinning to the original z-stream from that OCP release, or bumping it forward with CVE patches?

Because it's much less of maintenance burden to bump the z-stream to the latest than to hold it steady and backport CVEs to an older patch level.

Author: > in the OCP-version model, are we actually pinning to the original z-stream from that OCP release, or bumping it forward with CVE patches?

In OCP-version model we have the option - we can pin 3.2.19 on 5.0 and move to 3.2.latest on 5.1. Having versions based on OCP release makes this possible.

Just putting this on another words: are we going to bump haproxy 3.2 on 5.0.x as far as the latest one or stick on 3.2.19 and select patches/CVEs? Using OCP-version model is a way to replicate that same approach on the (y-1) of 5.1, and give the 5.1 version the freedom to bump to haproxy 3.2.latest.

Author: > No, we should fix those security vulnerabilities.

Rephrasing:

https://github.com/jcmoraisjr/openshift-enhancements/blob/91d29cfae59209b75052dbb786a1cda97048138d/enhancements/ingress/select-haproxy-version.md?plain=1#L652-L660

**Risk 3**: Dependency conflicts between HAProxy versions and their required
libraries (pcre, openssl, FIPS), including potential incompatibilities when
running older libraries from previous OCP releases on newer kernels.

**Mitigation**: Thoroughly test each supported version with its dependencies

> on each supported kernel version. Each HAProxy sidecar image is self-contained

Reviewer 2: Does "Thoroughly test each supported version with its dependencies on each supported kernel version" come down to testing on both RHEL 9 and RHEL 10?

Author: Part of this risk is mitigated by using the sidecar approach. The only sort of test that comes to mind is running a rhel9 based container on a rhel10 node and vice versa, wondering now if such sort of test fits here or this is instead already handled by rhel.

with all required dependencies, eliminating library isolation concerns.
Implement robust validation during version selection. Document tested
library/kernel combinations and known incompatibilities.

**Risk 4**: Complexity in troubleshooting when different IngressControllers
run different HAProxy versions.

**Mitigation**: Clearly expose the HAProxy version in IngressController
status. Add metrics and logging to identify which version is running.

> Include version information in support bundles.

Reviewer 2: Does "support bundles" refer to must-gather artifacts?

Author: Yes! Just updated.

### Drawbacks

This enhancement introduces additional complexity to the ingress subsystem:
- Increased maintenance burden for supporting multiple HAProxy versions
- Pod structure changes introducing an init container and sidecar container
- Inter-container communication via shared volume for configuration and
  admin socket
- More complex startup sequence with init container copying static files
- Additional testing required for version compatibility matrices

> - Potential for configuration drift across IngressControllers

Reviewer 2: It isn't clear what "configuration drift across IngressControllers" refers to.  The controller image is always the same for all ingresscontrollers, so it is running the same logic; are you referring to drift that could be introduced by the static files that the init container copies from the proxy image?

Author: This is about potentially running distinct HAProxy versions in the same cluster. Does it worth as a drawback, so just rephrase it?

matrix to a fixed number of versions. However, it does not provide an easy
or intuitive correlation to the OCP version currently running. Administrators
would need to consult documentation or status fields to understand which
actual HAProxy version or OCP release each identifier represents, making it
less transparent than the OCP-based referencing approach.

#### Alternative 3: Automatic Canary Testing

Implement automatic canary testing where the operator gradually rolls out
new HAProxy versions and monitors for issues.

**Why not selected**: While valuable, this is a more complex feature that
could be built on top of this enhancement in the future. The current
proposal provides the building blocks for manual canary testing by allowing
administrators to create separate IngressControllers with different
versions.

#### Alternative 4: Complete IngressController Image Selection

Allow selection of entire router images rather than just HAProxy versions.

**Why not selected**: This provides too much flexibility and could lead to
unsupported configurations. The goal is specifically to manage HAProxy
version risk while keeping other router components synchronized with the

> OpenShift release.

Reviewer 2: This would be nifty as an option under `unsupportedConfigOverrides`, but that would be separate from this EP.

Author: This one is about using the whole router instead, although the wording is a bit ambiguous. "Image" here means the router image.

Moving to https://github.com/openshift/enhancements/pull/1965#discussion_r3376524198

   clean separation of concerns, smaller individual images, and independent
   versioning while avoiding library isolation complexity.

2. What telemetry should be collected to track HAProxy version adoption and
   identify potential issues with specific versions?

3. ~~**HAProxy version pinning during initial feature adoption**: Since OCP 5.0
    is the first release where the `haproxyOCPVersion` field becomes available,
    how should administrators pin to OCP 4.22's HAProxy version when upgrading
    from OCP 4.22 to OCP 5.0?~~ **RESOLVED**: The migration should preserve the
    HAProxy version from 4.22 during upgrades from OCP 4.22 to OCP 5.0. The
    migration process will automatically populate the `haproxyOCPVersion` field with
    "OCP-4.22" for all existing IngressControllers during the upgrade to 5.0,
    ensuring HAProxy version stability. Administrators who want to adopt OCP
    5.0's newer HAProxy version must explicitly update the `haproxyOCPVersion` field
    to empty after the upgrade completes. This approach prioritizes upgrade
    safety and allows administrators to test and validate the new HAProxy version

>     on their own timeline.

Reviewer 2: Ah, this answers some of my earlier questions or comments about empty `haproxyOCPVersion`.  So on OpenShift 5.0, `haproxyOCPVersion: "OCP-4.22"` means HAProxy 2.8, and `haproxyOCPVersion: "OCP-5.0"` or empty/omitted `haproxyOCPVersion` means HAProxy 3.2, but we still default to HAProxy 2.8 (despite empty meaning HAProxy 3.2) due to this migration logic?  

This does raise some additional questions though: 

- What is the logic for determining that a particular IngressController resource should be migrated (or that it should not)?
- Is this migration going to be one-off for just OCP 4.22→5.0, or would we do something similar for every OCP 5._y_ → 5.(_y_+1) upgrade?
- Should we still use the terms "current" and "default" to describe the HAProxy version of a given OpenShift release, or do we need to abandon those terms to avoid confusion?

Reviewer 3: One more question - what are new installations for 5.0 defaulting to? 2.8 or 3.2? Is it a universal default for all clusters to 2.8 (upgraded and fresh installs)? We should be explicit about that.

Author: A few Q&A just to guide the changes I should made in EP.

> What is the logic for determining that a particular IngressController resource should be migrated

Do you mean the logic to automatically update an empty version? The main point is the inability for the user to define a version manually, since 4.22 does not have the API field, and a possible preference for the user to not have HAProxy update. So my first thought on this is:

* Populate the version field with `OCP-4.22` / `HAProxy-2.8` if empty and migration comes from `4.22` or older.

> Is this migration going to be one-off for just OCP 4.22→5.0

Yes, just because 4.22 does not have a way for the user to specify they want/don't want a HAProxy migration to happen.

> Should we still use the terms "current" and "default" to describe the HAProxy version

It was [suggested in the API review](https://github.com/openshift/enhancements/pull/1965#discussion_r3338213983) to not have defaulting value for the default behavior; missing or empty should do that. I don't have a strong opinion but tend to prefer not to use defaulting values as well. Thoughts on this?

> what are new installations for 5.0 defaulting to? 2.8 or 3.2?

3.2 because the version field should probably be empty, and default HAProxy version is always the latest one. The 4.y -> 5.y migration happens by overwriting the version field if empty and migration comes from 4.y.

4. **FIPS compliance and validation**: How do FIPS requirements impact the
   HAProxy sidecar container?
   - How is FIPS mode validated for each HAProxy version?
   - What are the certification implications of running FIPS-validated

>      libraries from older OCP releases on newer kernels?

Reviewer 2: FIPS is validated for RHEL.  From what I understand, the HAProxy RPM for a given combination of OpenShift release and HAProxy version will be built using the RHEL packages (including openssl-libs) for that OpenShift release, and it will be running on the RHEL kernel for that OpenShift release, so FIPS compliance will remain intact.

5. **Version to image mapping**: How does the operator maintain the mapping
   from `haproxyOCPVersion: "OCP-X.Y"` to the HAProxy sidecar container image
   reference?
   - Is the mapping hardcoded in the operator?
   - Stored in an API resource?

>    - When is this mapping validated (reconcile time vs cluster upgrade time)?

Reviewer 2: I figure it will come from `manifests/image-references`; cluster-ingress-operator from OpenShift 5.0 can use imagestream tags like "haproxy28-rhaos50" and "haproxy32-rhaos50".

Author: Ah good point! In fact this makes things easier for us. Can you confirm if this manifest or the data it provides is available in runtime? (well we can always embed it if not...)

6. **HAProxy version deprecation and continuous support window**: How should
   we handle HAProxy version support when OCP versions go end-of-life?
   - Example timeline:
     - Day 0: OCP 5.0 releases, supports HAProxy from 5.0 and 4.22
     - Day 90: OCP 5.1 releases, supports HAProxy from 5.1, 5.0, and 4.22
     - Day 180: OCP 4.21 goes EOL (but 4.22, 5.0, 5.1 are still supported)
   - On Day 180, if a cluster is running OCP 5.0 with `haproxyOCPVersion:
     "OCP-4.21"`, should the IngressController:
     - Continue running with OCP 4.21's HAProxy?
     - Set a warning/degraded condition?

>      - Be forcibly upgraded to a supported version?

Reviewer 2: Did you mean to write OCP 4.21, or OCP 4.22?  OCP 5.0 will never support OCP 4.21's HAProxy, will it?

If an IngressController resource on an OCP 5.0 cluster specifies "OCP-4.22", I figure we would continue to support the OCP 4.22 versoin of HAProxy (namely HAProxy 2.8) for the life of OCP 5.0.

Author: Ya I meant 4.22 instead.

Just answered it: haproxy is supported despite of the OCP having it as the default version went EOL.

   - Does the support matrix update when an OCP version goes EOL, even for
     clusters not yet upgraded to the next version?
   - How do we communicate the support lifecycle to administrators?

7. **Library compatibility across kernel versions**: What are the risks of
   running older dynamic libraries (pcre, openssl, FIPS modules from OCP 4.22)
   packaged in HAProxy sidecar images on a newer kernel (from OCP 5.1)?
   - Are there known incompatibilities between specific library/kernel
     version combinations?
   - How do we validate compatibility during testing?

>    - Should we document supported/tested combinations?

Reviewer 2: Is this still relevant?  I thought that we were planning to build release-specific RPMs.  That is, if OpenShift 5.0 supports HAProxy 2.8 and HAProxy 3.2, we will be building RPMs (1) with HAProxy 2.8 built for OpenShift 5.0 and (2) with HAProxy 3.2 built for OpenShift 5.0.

Author: Yeah this is already answered once we chose the sidecar approach. I'll update it as such.

   OCP release ships with all necessary HAProxy versions self-contained, without
   requiring cross-release dependencies or image reuse from previous OCP
   releases.

9. **HAProxy version correlation and visibility**: The IngressController
   status shows the OCP version (e.g., `status.effectiveHAProxyOCPVersion:
   "OCP-5.0"`) and HAProxy exposes its own version number in metrics.
   - Do administrators need an explicit mapping table (OCP version → HAProxy
     version number) in documentation?
   - Should the actual HAProxy version number appear in logs, events, or
     additional status fields?

10. **Version support criteria for 5.x releases with 4.x availability**: Once
    OCP 4.23 and newer 4.x versions start releasing (after OCP 5.0 and 5.x are
    available), what should be the criteria for supported HAProxy versions?
    - How does the "current and 2 previous minor versions" rule apply across
      major version boundaries?
    - Should 5.x releases support HAProxy versions from 4.x releases, or only
      from 5.x releases?
    - Example: If OCP 5.2 and OCP 4.25 are both actively supported, should
      OCP 5.2 allow `haproxyOCPVersion: "OCP-4.25"`?

>     - Should there be separate support tracks for 4.x and 5.x?

Reviewer 2: Yes, this does get tricky.  It would make sense to me for "previous" to include only versions from which upgrades were possible.  From what I understand:

- To 5.0, it will be possible to upgrade from 4.22.
- To 5.1, it will be possible to upgrade from 4.23 or 5.0.
- To 5.2, it will be possible to upgrade from 5.1.
- To 5.3, it will be possible to upgrade from 5.2.
- And so on.

So "current and 2 previous minor versions" would mean the following:

- For 5.0, allowed values should be "OCP-5.0" (_n_) and "OCP-4.22" (_n_-1).  As an exception, we are saying we do not allow _n_-2 for 5.0.
- For 5.1, allowed values should be "OCP-5.1" (_n_), "OCP-5.0" (_n_-1), "OCP-4.23" (also _n_-1), and "OCP-4.22" (_n_-2 where _n_-1 = 5.0, and also _n_-2 = where _n_-1 = 4.23).
- For 5.2, allowed values should be "OCP-5.2" (_n_), "OCP-5.1" (_n_-1), "OCP-4.23" (_n_-2), and "OCP-5.0" (also _n_-2).
- For 5.3, allowed values should be "OCP-5.3" (_n_), "OCP-5.2" (_n_-1), and "OCP-5.1" (_n_-2).  From here the pattern becomes simple.
- For 5.4, allowed values should be "OCP-5.4" (_n_), "OCP-5.3" (_n_-1), and "OCP-5.2" (_n_-2).
- For 5.5, allowed values should be "OCP-5.5" (_n_), "OCP-5.4" (_n_-1), and "OCP-5.3" (_n_-2).
- ...
- For 6.0, who knows.  ¯\\\_(ツ)\_/¯

So supporting _n_-2 actually means up to 4 versions.  

But see also https://github.com/openshift/enhancements/pull/1965#discussion_r3353400629 where I point out that we might actually need _n_-3 if the goal is to bridge extended-support releases.

Author: In the following scenario:

* User moves from 4.22 to 4.24
* Than user moves from 4.24 to 5.1

Should 5.1 support 4.24 compatibility somehow? Or instead, any 4.y where y>=23 should pin haproxy, so `OCP-4.22` / `HAProxy-2.8` from OCP 5.1 means also compatibility with 4.24 as well? Does it make sense to name the version as `OCP-4` (dropping the `.22` suffix) in case we use the OCP-version model?

Reviewer 3: BTW I think 4.23 is the end of the OCP 4 releases. Hopefully that helps the confusion (we should remove mention of 4.24+).

I think a matrix/table/chart/diagram/visual showing the supported version between 4.22, 4.23, 5.0, 5.1 would be helpful in the EP somewhere. The straddling of 4.22 and 4.23 with 5.0 and 5.1 sounds a bit complicated.

Author: This makes things easier; just answered open question 10 since 4.x won't be a fix branch of 5.x.

11. **Custom HAProxy image configuration**: Should administrators be allowed to
    configure their own HAProxy image in an unsupported fashion, instead of
    always relying on the OCP-related ones?
    - Could this be done via a documented but unsupported API field that is
      officially exposed but carries no support guarantees?
    - Alternatively, should this be implemented as a hidden and undocumented
      field that technically works but is not publicly exposed in the API
      schema or documentation?
    - What are the implications for support, security, and FIPS compliance when
      allowing custom images?
    - How does this interact with the supported OCP version mappings?

>     - What validations should be applied when a custom image is selected?

Reviewer 2: This seems a lot like Alternative 4, right?  And as I said there (see https://github.com/openshift/enhancements/pull/1965#discussion_r3353583774), this would make sense as an option under `unsupportedConfigOverrides`, but it wouldn't necessarily be part of this enhancement.

Author: Not exactly. Alternative 4 is grabbing the whole router from an older OCP release instead of just running a distinct haproxy as a sidecar. Really not a good option.

This one is simply allowing the user to run their own haproxy image, either from a public hub or manually building themselves and pushing to a hub visible to the ocp cluster.

> This would be nifty as an option under `unsupportedConfigOverrides`, but that would be separate from this EP.

Do you mean having another EP for that, or providing this support without mentioning in this EP - in case this is a valid addition of course.

## Test Plan

The test plan for this enhancement must include:

**Unit Tests**:
- API validation for the `haproxyOCPVersion` field
- Version selection logic in the operator
- Version compatibility validation
- Upgrade scenario handling

**Integration Tests**:
- IngressController creation with different HAProxy versions
- Switching between HAProxy versions on existing IngressControllers
- Behavior during cluster upgrades with various version configurations
- Validation of version limits (maximum 3 distinct versions)

**E2E Tests**:
All E2E tests must include the `[sig-network-edge]` and the
`[OCPFeatureGate:SelectableHAProxyVersion]` labels for the component, and
appropriate test type labels like `[Suite:openshift/conformance/parallel]`,
`[Serial]`, `[Slow]`, or `[Disruptive]` as needed.

Tests must cover:
- Basic functionality: using empty value (default) and specific OpenShift versions

> - Version persistence across cluster upgrades

Reviewer 2: Is version persistence specific to 4.22→5.0, or will we keep it for 5._y_ → 5.(_y_+1)?

Author: Both, just changed this into two distinct items, wdyt:

```
- Version persistence across cluster upgrades, using API field on OCP 5.0 or newer
- Version persistence when migrating from 4.22 to 5.0, where API field is not available
```

- Multiple IngressControllers with different HAProxy versions
- Validation that routes work correctly with different HAProxy versions

> - Performance and resource consumption with multiple versions

Reviewer 2: Is this a test scenario we need to plan with the PerfScale team, or would it be an Origin E2E test?

Author: I'm not sure. We already agreed on other thread that the change in resource consumption is negligible, is it the same for performance? Should the regular performance tests be enough? If so maybe this item can be dropped.

- HAProxy configuration reload functionality via admin socket API
- Init container startup and static file copying

> - Sidecar container communication and shared volume functionality

Reviewer 2: Are these covered by existing tests, or do we need something more specific to verify outward-facing behavior or internal mechanics?

Author: In fact it is explored all the time - all the communication between router and haproxy happens via the shared volume. It is here more like something that changed, not exactly as an action to be taken. That said, does it still fit here?

**Negative Tests**:
- Attempting to use unsupported/unavailable versions
- Invalid version format strings
- Verify `Upgradeable=False` condition when attempting to upgrade to an
  OpenShift version that does not support the currently selected

>   `haproxyOCPVersion`

Reviewer 2: I'm guessing this would be a unit test?  The wording is a bit strange; maybe this is clearer?
```suggestion
- Verify the operator sets `Upgradeable=False` when `haproxyOCPVersion` specifies
  a version that is too old to be supported by the next OpenShift minor release.
```

Author: Good point. Rephrased and moved to unit tests.

## Graduation Criteria

**Testing Requirements**:
- Minimum 5 tests per feature gate
- All tests must run at least 7 times per week
- All tests must run at least 14 times per supported platform
- Tests must be in place at least 14 days before branch cut
- All tests must pass at least 95% of the time
- Tests running on all supported platforms: AWS (HA/Single), Azure (HA),

>   GCP (HA), vSphere (HA), Baremetal (HA with IPv4/IPv6/Dual)

Reviewer 2: This seems like a lot of boilerplate?

Author: Claude suggestion, just left it here for a wider discussion.

Just changed to this instead, wdyt:

```
We will introduce the feature as TechPreviewNoUpgrade during the OpenShift 5.0
development, with the goal of graduating it to GA before it is released.
```

### Dev Preview -> Tech Preview

> - Ability to select and use different HAProxy versions end-to-end

Reviewer 2: What does "end-to-end" mean here?  Is this supposed to be about end-to-end testing?

> - End user documentation covering use cases and migration strategies

Reviewer 2: A lot of these criteria look like boilerplate, but would we really need to have end-user documentation for graduation to TPNU?

Author: My bad, we're not using dev preview, just dropped all the section.

Regarding the items, they are a sort of boiler plate that came from the template.

- API stability with no planned breaking changes
- Sufficient test coverage across supported platforms
- Gather feedback from early adopters
- Metrics exposed for HAProxy version tracking
- Alerts defined for version compatibility issues

### Tech Preview -> GA

> - Extensive testing including upgrade and scale scenarios

Reviewer 2: What kind of scale scenarios?

> - At least one full release cycle in Tech Preview

Reviewer 2: Does that mean it would ship as TPNU in 5.0?

Author: There were a number of reviews on most of the items, left only this discussion open providing the result of a refactor, closing all the other threads.

Old content was the result of parsing the template and some particularities of the EP, mostly done by Claude, and I shared here for a wider discussion.

New content is:

```
### Tech Preview -> GA

- Payload tests pass with HAProxy selection enabled
- All known regressions are fixed
- Extensive testing including upgrade scenarios
- End-to-end tests included in standard conformance suites
- Support procedures documented for troubleshooting version issues
- Limitations, such as migration criterias, are documented
```

How does it look now?

> - Available by default with empty `haproxyOCPVersion` (current release) behavior

Reviewer 2: What does this mean?

- Telemetry data showing feature adoption and stability
- User-facing documentation in openshift-docs

> - Performance testing showing no regression

Reviewer 2: We don't usually have telemetry, user-facing documentation, or performance testing done before graduating a feature to GA.

- Support procedures documented for troubleshooting version issues
- End-to-end tests included in standard conformance suites

### Removing a deprecated feature

N/A - This is a new feature.

## Upgrade / Downgrade Strategy

**Upgrades**:

When upgrading an OpenShift cluster:
1. IngressControllers with `haproxyOCPVersion` unset (empty) will automatically

>    use the new HAProxy version from the upgraded OpenShift release.

Reviewer 2: ```suggestion
1. IngressControllers with `haproxyOCPVersion` unset (empty) will automatically
   use the new HAProxy version from the upgraded OpenShift release, except in the
   case of OpenShift 5.0 (see below). 
```

upgrades. The default behavior (using the latest HAProxy) is preserved.

**Special Case - Upgrading to OCP 5.0 (First Release with Feature)**:

When upgrading from OCP 4.22 (or earlier) to OCP 5.0, the `haproxyOCPVersion`
field does not exist in the source release. The behavior during this
bootstrapping scenario is addressed in Open Question #3: the migration process
will automatically populate the `haproxyOCPVersion` field with "OCP-4.22" for all
existing IngressControllers during the upgrade to 5.0.

## Version Skew Strategy

This enhancement handles version skew through the following mechanisms:

**Control Plane / Data Plane Skew**:
The ingress-controller-operator (control plane) must support serving
multiple HAProxy versions to router pods (data plane). The operator version
determines which HAProxy versions are available, not the router pod version.

**Multi-Version Support Window**:
The operator maintains compatibility with HAProxy versions from the current
OpenShift release and up to 2 previous releases. During a rolling upgrade,
different router pods may temporarily run different HAProxy versions, which

> is acceptable as each IngressController operates independently.

Reviewer 2: It isn't clear how the fact that different router deployments ("each IngressController") can now run different HAProxy versions is related to the fact that a rolling update can cause a single deployment to have pods running different HAProxy versions.  Anyway, the situation with rolling deployments isn't any different for this enhancement than it is for any other enhancement, is it?

Author: Good catch, I'm not sure how this ended up this way. I intended to state about HAProxys from distinct pods operating independently, but as you said, this is just a vanilla rolling update. The only particularity is about HAProxy version potentially changing, which should be acceptable, so I simply dropped the last part of the last sentence. wdyt?

### SLIs for API Extensions

This enhancement modifies an existing CRD (IngressController) but does not
add webhooks or aggregated API servers. The existing SLIs for
IngressController resources apply:

- `cluster-ingress-operator` condition `Available=True`
- `cluster-ingress-operator` condition `Degraded=False`
- IngressController resource status conditions

### Impact on Existing SLIs

**API Throughput**: Minimal impact. The new field is optional and adds
negligible processing overhead during IngressController reconciliation.

**Scalability**: No significant impact. The operator already manages
multiple IngressControllers, and version selection is a one-time decision
during reconciliation.

**API Availability**: No impact. The enhancement does not add API
dependencies or external calls.

### Measurement and Testing

Performance impact will be measured through:
- Existing OpenShift scalability tests that create and manage multiple
  IngressControllers
- Component Readiness regression testing across all supported platforms
- Monitoring IngressController reconciliation times with version selection
  enabled

QE will measure these metrics in standard CI runs for each supported

> platform.

Reviewer 2: Out of curiosity, where do these sections come from?  I don't see these sections in the template, but I do see a few recent EPs have "Impact on Existing SLIs".

Author: Claude helped me with this EP, which should be based on the template, and I didn't realize it was not part of it. It seems we're not loosing anything here, right? If so I'll remove it.

Author: Just revisited the section, in fact we're not extending API not adding webhook, so I simply dropped this section.

## Support Procedures

* **Failure Mode 1**: Requested HAProxy version is not available
   * **Impact**: IngressController enters degraded state, uses fallback version
or refuses to reconcile.
   * **Detection**: IngressController status condition `Degraded=True` with
reason `UnavailableHAProxyVersion`. Operator logs include details about
missing version.

>    * **Teams**: Networking team (ingress maintainers) handles escalations.

Reviewer 2: How can this happen?  Does this imply that the release payload is missing one of the images that are part of the release, or that the validation somehow allowed an invalid version to be specified?

Author: This should only happen if the user has the ability to provide a HAProxy image by themselves. Since this should not be part of this EP, wondering if I can simply drop this failure mode.

* **Failure Mode 2**: Dependency mismatch between HAProxy and libraries
   * **Impact**: Router pods fail to start or crash at runtime.
   * **Detection**: Router pod crash loops, readiness probe failures, operator
degraded condition. Logs show library loading errors.

>    * **Teams**: Networking team (ingress maintainers) handles escalations.

Reviewer 2: How can this happen?  Does this imply we somehow manage to build a container image with missing libraries and ship it?  A lot of things need to fail inexplicably for this to happen—Yum's dependency checks, numerous E2E tests and CI jobs, QA...

Hm, maybe this is a holdover from one of the alternative implementations that change the linker/loader configuration?

Author: Exactly, this might happen in other implementation alternatives. Explain this in this failure mode or simply drop it since it does not apply in the approach we chose?

```go
// HAProxyOCPVersion specifies the OCP release whose default HAProxy version
// should be applied to this IngressController.
//
// For example, selecting "4.22" will configure the IngressController to use 
// the default HAProxy version shipped with OCP 4.22.
//
// Valid values depend on the current OpenShift release:
// - Empty string (default): Use the default HAProxy version from the current
//   OpenShift release (the most recent version available)
// - "X.Y": Use the default HAProxy version from OpenShift release X.Y
//
// For ELC (Extended Lifecycle) releases, three values are supported: empty,
// the current version, or the previous ELC version. Empty and the current
// version use the same HAProxy image; specifying the current version
// explicitly pins the version during cluster upgrades.
//
// For non-ELC releases, two values are supported: empty or the previous ELC
// version.
//
// If a specific OCP version is set and would become unsupported in a target
// cluster upgrade, the ingress ClusterOperator's Upgradeable condition will
// be set to False, blocking the cluster upgrade until this field is updated
// to empty or a supported version.
//
// +optional

> // +kubebuilder:validation:MinLength=0

Reviewer: I think this conflicts with your regex pattern below, specifically if the user tries to set an empty string `""`.

Based on the godoc I think there's no difference between unset and empty, so since the field is optional, you can probably just validate against valid OCP releases, and tell the user to unset this field if they do not have an opinion.

You could also make this a `type OCPVersion string` and validate there, instead of redefining for spec and status.

Author: Not accepting empty is a valid approach to me, we ended up a bit ambiguous in the EP regarding that.

@Miciah @gcs278 do we have any possible scenario where we should accept empty as a valid input? Does it make sense to review all the EP and drop empty as an option instead?

> You could also make this a `type OCPVersion string` and validate there

Sounds good, I'll update it accordingly.

Reviewer 3: I don't think there is any scenario or benefit to allowing `""` explicitly. It's not a pointer, so `""` is the same as unset. I think it's better to disallow `""` to be more restrictive. If we ever need convert to a pointer and distinguish between unset and `""` in the future, allowing it now would make it more complicated.

Author: Both addressed: validation does not allow empty anymore; created a separated type OCPVersion with the validations.

// +kubebuilder:validation:MaxLength=8
// +kubebuilder:validation:Pattern=`^[0-9]+\.[0-9]+$`
// +openshift:enable:FeatureGate=IngressControllerMultipleHAProxyVersions
HAProxyOCPVersion string `json:"haproxyOCPVersion,omitempty"`
```

The proposed API fields for the IngressController status:

```go
// EffectiveHAProxyOCPVersion reports the OCP release whose default HAProxy

> // version is currently in use by this IngressController. This reflects the

Reviewer: Do you expect this to always be set after the controller has synced at least once?

Author: Yes I do expect, but having a closer look on the status I'm wondering how much value does this have the way it is, and how much it is aligned with other similar status from the OpenShift API. I might be missing some details here, but we do expect to always fill the same version the user provided, so having it filled is more or less the same of having some condition being flagged to true.

version for the current OpenShift release and a pinned version from the
previous Extended Lifecycle (ELC) release. ELC releases occur every 3 minor
versions (e.g., 5.2, 5.5, 5.8). Each OCP version has a 1:1 mapping to a
specific HAProxy version. The feature begins with OCP 5.0.

**Version Selection Rules**:
- **ELC releases** (e.g., 5.2, 5.5, 5.8): Support unset (field not specified),
  the current version (e.g., "5.2"), or the previous ELC version (e.g., "4.22").
  Note that unset and the current version both use the same HAProxy image;
  specifying the current version explicitly pins the version during cluster
  upgrades.
- **Non-ELC releases** (e.g., 5.0, 5.1, 5.3): Support unset (field not specified)
  or the previous ELC version only (e.g., "4.22" for OCP 5.0-5.1, "5.2" for
  OCP 5.3-5.4).

**Supported Version Matrix**:

The following table shows which `haproxyOCPVersion` field values are valid for
each OpenShift release, and which HAProxy images are shipped with each release:

| OCP Release | Release Type | Valid Field Values | HAProxy Images Shipped | Notes |
|-------------|--------------|-------------------|------------------------|-------|

> | 4.22 | ELC | N/A | N/A | Feature not available; auto-migrated to "4.22" when upgrading to 5.0 |

Reviewer 3: *nit* it's EUS on 4, ELC on 5.0.

Also - what about 4.23? 4.23 == 5.0 right? So my assumption is that we are also pinning to HAproxy 2.8 (OCP-4.22) for 4.23 as well. 

```suggestion
| 4.22 | EUS | N/A | N/A | Feature not available; auto-migrated to "4.22" when upgrading to 4.23 or 5.0 |
```

> | 5.0 | Non-ELC | unset<br/>`"4.22"` | `ose-haproxy:5.0`<br/>`ose-haproxy:4.22` | First release with this feature |

Reviewer 3: Could we add an explicit `4.23`? Is this what you had in mind? I think the assumption is that it will be available on 4.23 based on some [threads](https://redhat-internal.slack.com/archives/GNR5FB2M6/p1779208047803919?thread_ts=1778678338.907069&cid=GNR5FB2M6) I found.

```suggestion
| 4.23 | Non-EUS | unset<br/>`"4.22"` | `ose-haproxy:4.23`<br/>`ose-haproxy:4.22` | Functionally equivalent to 5.0; first 4.x release with this feature |
| 5.0 | Non-ELC | unset<br/>`"4.22"` | `ose-haproxy:5.0`<br/>`ose-haproxy:4.22` | First release with this feature |
```

| 5.1 | Non-ELC | unset<br/>`"4.22"` | `ose-haproxy:5.1`<br/>`ose-haproxy:4.22` | |

> | 5.2 | **ELC** | unset<br/>`"5.2"`<br/>`"4.22"` | `ose-haproxy:5.2`<br/>`ose-haproxy:4.22` | Unset and "5.2" use same image; "5.2" pins during upgrades |

Reviewer 3: Might be worth adding upgradeable=false notes for each version that needs it.

```suggestion
| 5.2 | **ELC** | unset<br/>`"5.2"`<br/>`"4.22"` | `ose-haproxy:5.2`<br/>`ose-haproxy:4.22` | Unset and "5.2" use same image; "5.2" pins during upgrades. First release where `Upgradeable=False` applies: upgrading to 5.3 with `"4.22"` is blocked. |
```

| 5.3 | Non-ELC | unset<br/>`"5.2"` | `ose-haproxy:5.3`<br/>`ose-haproxy:5.2` | "4.22" no longer valid; 5.2 is now the previous ELC |
| 5.4 | Non-ELC | unset<br/>`"5.2"` | `ose-haproxy:5.4`<br/>`ose-haproxy:5.2` | |

> | 5.5 | **ELC** | unset<br/>`"5.5"`<br/>`"5.2"` | `ose-haproxy:5.5`<br/>`ose-haproxy:5.2` | Unset and "5.5" use same image; "5.5" pins during upgrades |

Reviewer 3: ````suggestion
| 5.5 | **ELC** | unset<br/>`"5.5"`<br/>`"5.2"` | `ose-haproxy:5.5`<br/>`ose-haproxy:5.2` | Unset and "5.5" use same image; "5.5" pins during upgrades. Upgrading to 5.6 with `"5.2"` is blocked by `Upgradeable=False`. |
````

| 5.6 | Non-ELC | unset<br/>`"5.5"` | `ose-haproxy:5.6`<br/>`ose-haproxy:5.5` | "5.2" no longer valid; 5.5 is now the previous ELC |
| 5.7 | Non-ELC | unset<br/>`"5.5"` | `ose-haproxy:5.7`<br/>`ose-haproxy:5.5` | |

> | 5.8 | **ELC** | unset<br/>`"5.8"`<br/>`"5.5"` | `ose-haproxy:5.8`<br/>`ose-haproxy:5.5` | Unset and "5.8" use same image; "5.8" pins during upgrades |

Reviewer 3: ````suggestion
| 5.8 | **ELC** | unset<br/>`"5.8"`<br/>`"5.5"` | `ose-haproxy:5.8`<br/>`ose-haproxy:5.5` | Unset and "5.8" use same image; "5.8" pins during upgrades. Upgrading to 5.9 with `"5.5"` is blocked by `Upgradeable=False`. |
````

builds for each affected version. Clearly document supported versions and
deprecation timelines. Provide warnings in the IngressController status when
using older versions. Consider implementing alerts when versions reach
end-of-support.

**Risk 3**: Dependency conflicts between HAProxy versions and their required
libraries (pcre, openssl, FIPS), including potential incompatibilities when
running older libraries from previous OCP releases on newer kernels.

**Mitigation**: Thoroughly test each supported version with its dependencies
on each supported kernel version. Each HAProxy sidecar image is self-contained
with all required dependencies, eliminating library isolation concerns.
Implement robust validation during version selection. Document tested
library/kernel combinations and known incompatibilities.

**Risk 4**: Complexity in troubleshooting when different IngressControllers
run different HAProxy versions.

**Mitigation**: Clearly expose the HAProxy version in IngressController
status. Add metrics and logging to identify which version is running.
Include version information in the must-gather artifacts.

> 

Reviewer 3: It's worth a risk and mitigation to explain that this feature motivates us to bump HAProxy minor versions at every ELC release, and that bumping too much can reduce the value of this feature.

````suggestion

**Risk 5**: HAProxy minor version bump cadence misalignment with ELC releases.
If HAProxy minor versions are bumped more than once per ELC window, the gap
between the previous ELC and the current version grows beyond one LTS
generation. For example, if 5.2 (ELC) ships HAProxy 3.2, 5.3 bumps to 3.4,
and 5.5 (ELC) bumps again to 3.6, then pinning to "5.2" on 5.5 skips two
HAProxy LTS generations instead of one.

**Mitigation**: Align HAProxy minor version bumps with ELC releases — each
ELC release adopts the next HAProxy LTS, and non-ELC releases only carry
z-stream patches. HAProxy LTS releases occur approximately once per year,
matching the ELC cadence naturally. Bumping once between ELC releases is
acceptable since pinned users are unaffected, but the default policy should
be one HAProxy LTS bump per ELC cycle.
````

Author: I added as suggested, but I'm missing in the description why bump misalignment would be a problem. e.g. we're going from 2.8 to 3.2 in OCP5.0. Also, in case HAProxy LTS wouldn't be once per year, just like OpenShift ELC, we'd eventually need to have some misalignment during the OCP releases.

Reviewer 3: Yea, good point. We are indeed skipping 3.0 this release and jumping 2 minor versions. So this is risk is probably more of an observation rather than a hard risk, but my point here was: jumping two minor versions is inherently riskier than one for users.

Feel free to drop it, or reframe it as a non-risk/observation in a different part of the EP. I mainly wanted to think through the implications of release misalignment with HAProxy and OCP.

Author: Nice, just added another implementation detail talking about the proposed approach.

Allow selection of entire router images rather than just HAProxy versions.

**Why not selected**: This provides too much flexibility and could lead to
unsupported configurations. The goal is specifically to manage HAProxy
version risk while keeping other router components synchronized with the
OpenShift release.

## Open Questions [optional]

1. ~~Which implementation proposal (1, 2, or 3) should be selected for the
   initial implementation?~~ **RESOLVED**: Proposal 2 (External HAProxy Images
   with Sidecar Deployment) has been selected. The sidecar approach provides
   clean separation of concerns, smaller individual images, and independent
   versioning while avoiding library isolation complexity.

2. What telemetry should be collected to track HAProxy version adoption and
   identify potential issues with specific versions?

3. ~~**HAProxy version pinning during initial feature adoption**: Since OCP 5.0
    is the first release where the `haproxyOCPVersion` field becomes available,
    how should administrators pin to OCP 4.22's HAProxy version when upgrading
    from OCP 4.22 to OCP 5.0?~~ **RESOLVED**: The migration should preserve the

>     HAProxy version from 4.22 during upgrades from OCP 4.22 to OCP 5.0. The

Reviewer 3: ```suggestion
    HAProxy version from 4.22 during upgrades from OCP 4.22 to OCP 4.23 and OCP 5.0. The
```

  Note that unset and the current version both use the same HAProxy image;
  specifying the current version explicitly pins the version during cluster
  upgrades.
- **Non-ELC releases** (e.g., 5.0, 5.1, 5.3): Support unset (field not specified)
  or the previous ELC version only (e.g., "4.22" for OCP 5.0-5.1, "5.2" for
  OCP 5.3-5.4).

**Supported Version Matrix**:

The following table shows which `haproxyOCPVersion` field values are valid for
each OpenShift release, and which HAProxy images are shipped with each release:

| OCP Release | Release Type | Valid Field Values | HAProxy Images Shipped | Notes |
|-------------|--------------|-------------------|------------------------|-------|
| 4.22 | EUS | N/A | N/A | Feature not available; auto-migrated to "4.22" when upgrading to 4.23 or 5.0 |
| 4.23 | Non-EUS | unset<br/>`"4.22"` | `ose-haproxy:4.23`<br/>`ose-haproxy:4.22` | Functionally equivalent to 5.0; first 4.x release with this feature |
| 5.0 | Non-ELC | unset<br/>`"4.22"` | `ose-haproxy:5.0`<br/>`ose-haproxy:4.22` | First release with this feature |
| 5.1 | Non-ELC | unset<br/>`"4.22"` | `ose-haproxy:5.1`<br/>`ose-haproxy:4.22` | |

> | 5.2 | **ELC** | unset<br/>`"5.2"`<br/>`"4.22"` | `ose-haproxy:5.2`<br/>`ose-haproxy:4.22` | Unset and "5.2" use same image; "5.2" pins during upgrades. First release where `Upgradeable=False` applies: upgrading to 5.3 with `"4.22"` is blocked. |

Reviewer 3: This table is really helpful to visualize the design, thanks!

There is a pretty concerning EOL problem here though - with this plan, HAProxy 2.8 is still supported until 5.2. HAProxy 2.8 is EOL in in Feb 2028. And the EP says:

> For example, if OCP 4.22 goes EOL but OCP 5.2 still supports "4.22" as a selectable HAProxy version (as the previous ELC), Red Hat will continue to maintain and apply security fixes to that HAProxy version. Support ends only when the HAProxy version is no longer referenced by any supported OCP release.

So we'd be maintaining HAProxy 2.8 for potentially >4 years past its upstream EOL on a long-lived OCP release? We're already doing this on 4.22 for about 1 year, so it's not a new problem — but 4x the post-EOL maintenance window is a significant increase.

The only other option is to caveat this previous HAProxy verison support with: previous pinned HAProxy versions are only supported until they go EOL upstream (then surface a `HAProxyVersionDeprecated=True` condition).

However, that's messy for users on long-term support having to do a mid-z-stream HAProxy upgrade to continue getting CVE fixes. But I guess they have to upgrade at some point 🤷‍♂️

It's worth a discussion. Can we carry HAProxy 2.8 for 4 years past it's EOL time? How confident do you feel with achieving that?

Reviewer 3: Another option: give HAProxy 2.8 a shorter migration window as a one-off. Drop "4.22" support from 5.0 --> 5.1 instead of carrying it all of the way to 5.3. It breaks the ELC-to-ELC migration pattern, but it still gives users one release cycle on 5.0 to test and migrate.

Maybe that's better than committing to 4 years of post-EOL maintenance? It might be justified as a one-off solution to push users on to HAProxy 3.2 quicker.

Author: Updated doc and table, now 2.8 is dropped just after 5.0

Allow selection of entire router images rather than just HAProxy versions.

**Why not selected**: This provides too much flexibility and could lead to
unsupported configurations. The goal is specifically to manage HAProxy
version risk while keeping other router components synchronized with the
OpenShift release.

## Open Questions [optional]

1. ~~Which implementation proposal (1, 2, or 3) should be selected for the
   initial implementation?~~ **RESOLVED**: Proposal 2 (External HAProxy Images
   with Sidecar Deployment) has been selected. The sidecar approach provides
   clean separation of concerns, smaller individual images, and independent
   versioning while avoiding library isolation complexity.

2. What telemetry should be collected to track HAProxy version adoption and
   identify potential issues with specific versions?

3. ~~**HAProxy version pinning during initial feature adoption**: Since OCP 5.0
    is the first release where the `haproxyOCPVersion` field becomes available,
    how should administrators pin to OCP 4.22's HAProxy version when upgrading
    from OCP 4.22 to OCP 5.0?~~ **RESOLVED**: The migration should preserve the

>     HAProxy version from 4.22 during upgrades from OCP 4.22 to OCP 4.23 or OCP

Reviewer 3: FYI slack [thread](https://redhat-internal.slack.com/archives/CCH60A77E/p1781646850933419) is discussing this.

Trevor (from the upgrades team) and Ben are pushing for float-by-default. I think they have reasonable points, but it is arguably more risky than pin-by-default. I do agree that, with pin-by-default, requiring every cluster admin to bump every IngressController when HAProxy 2.8 is no longer supported is not a great design. Their argument is that we should default to upgrading, and the exception should be rolling back to HAProxy 2.8.

Both of them are strongly against different defaults for upgrade vs. fresh installs.

Ultimately, I think we should consider following their opinion, but wait for feedback from @Miciah. Either way, it'd be another good thing to document this decision in this EP (if float-by-default, then pin-by-default could be an alternative or visa versa).

Author: As we discussed on Slack:

* Moving to float-by-default approach, using the backported api on 4.22 as the placeholder of the opt-in to pin haproxy 2.8 / OCP 4.22.
* haproxy 2.8 / OCP 4.22 mode is only supported until 5.0, dropping its support on 5.1
* Redesign of the version transition pattern in a way it's not coupled with OCP lifecycle.

the field is unset (the default), it always uses the default HAProxy version
from the current OpenShift release.

When an administrator specifies an OpenShift release version, the
IngressController will use the HAProxy version that shipped with that
specific OpenShift release. The cluster-ingress-operator will manage the
deployment of the appropriate HAProxy binary and its dependencies (pcre,
openssl, FIPS libraries) to support the selected version.

Exactly 2 distinct HAProxy versions are supported simultaneously: the default
version for the current OpenShift release and a pinned version from a previous
supported release. Each OCP version has a 1:1 mapping to a specific HAProxy
version. The feature begins with OCP 5.0.

**Version Selection Rules**:

Each OpenShift release supports exactly 2 distinct HAProxy versions: the default
version for the current release and a pinned version from a previous supported
release. The supported values for `haproxyVersion` are:

- **OCP 5.0**: Supports unset and "2.8"
  - Unset uses HAProxy 3.2 (default for 5.0)
  - "2.8" uses HAProxy 2.8 from the 4.22 release
  
- **OCP 5.1**: Supports unset and "3.2"
  - Unset uses HAProxy 3.2 (default for 5.1)

>   - "3.2" uses HAProxy 3.2 from the 5.0 release stream (pins to 5.0 z-stream patches)

Reviewer 3: ```suggestion
  - "3.2" uses the latest HAProxy 3.2.z available for this OpenShift release (pins the minor version; z-stream floats with each OCP release)
```

  - Note: "2.8" is dropped; HAProxy 2.8 support ends
  

> - **OCP 5.2**: Supports unset, "3.4", and "3.2"

Reviewer: These updated values look right to me, thanks!

Could you also detail out 5.3, 5.4 and 5.5, since 5.2 and 5.5 will be EUS, just so we can visualize what the EUS-to-EUS skip level would look like if we had to support that skip-level?

Author: Sure, added 5.3, 5.4, 5.5, and also the API-only backport 4.22.

  - Unset uses HAProxy 3.4 (default for 5.2)
  - "3.4" uses HAProxy 3.4 (explicitly pins version during cluster upgrades)

>   - "3.2" uses HAProxy 3.2 from the 5.0 release stream

Reviewer 3: ```suggestion
  - "3.2" uses the latest HAProxy 3.2.z available for this OpenShift release
```

  - Note: Unset and "3.4" use the same HAProxy image; specifying "3.4" explicitly
    pins the version during cluster upgrades

**Supported Version Matrix**:

The following table shows which `haproxyVersion` field values are valid for
each OpenShift release, and which HAProxy images are shipped with each release:

| OCP Release | Valid Field Values | HAProxy Images Shipped | Notes |
|-------------|-------------------|------------------------|-------|
| 4.22 | unset<br/>`"2.8"` | N/A | **API-only backport**: Field available but no operator implementation. Setting `"2.8"` allows pinning HAProxy 2.8 when upgrading to 5.0. |
| 5.0 | unset<br/>`"3.2"`<br/>`"2.8"` | `ose-haproxy:5.0` (3.2)<br/>`ose-haproxy:4.22` (2.8) | OCP 5.0 introduces HAProxy 3.2. Supports "3.2" (HAProxy 3.2) and "2.8" (HAProxy 2.8). |
| 5.1 | unset<br/>`"3.2"` | `ose-haproxy:5.1` (3.2.z)<br/>`ose-haproxy:5.0` (3.2.z) | HAProxy 2.8 dropped. Both values use HAProxy 3.2 but from different release streams. Upgrading from 5.0 with `"2.8"` is blocked. |

> | 5.2 | unset<br/>`"3.4"`<br/>`"3.2"` | `ose-haproxy:5.2` (3.4)<br/>`ose-haproxy:5.0` (3.2.z) | OCP 5.2 introduces HAProxy 3.4. Unset/"3.4" use HAProxy 3.4. "3.2" continues as pinnable version, using HAProxy 3.2 from 5.0 release stream. Unset and "3.4" use the same image; "3.4" pins during upgrades. |

Reviewer 3: Here's a suggestion for the new y-stream pinned z-stream floating. 

Interestingly, on 5.1, there is no previous HAProxy version to carry. 2.8 was dropped and no new minor was introduced, so unset and "3.2" resolve to the same thing.

Also, the HAProxy Images Shipped column doesn't add much with y-stream floating. Since the z-stream isn't frozen, the images are always built for the current OCP release rather than carried from a previous one. The valid field values already tell the user everything they need to know.

```suggestion
| OCP Release | Valid Field Values | HAProxy Versions Available | Notes |
|-------------|-------------------|---------------------------|-------|
| 4.22 | unset<br/>`"2.8"` | 2.8 | **API-only backport**: Field available but no operator implementation. Setting `"2.8"` allows pinning HAProxy 2.8 when upgrading to 5.0. |
| 5.0 | unset<br/>`"3.2"`<br/>`"2.8"` | 3.2, 2.8 | OCP 5.0 introduces HAProxy 3.2. Supports "3.2" (HAProxy 3.2) and "2.8" (HAProxy 2.8). |
| 5.1 | unset<br/>`"3.2"` | 3.2 | HAProxy 2.8 dropped. Both unset and "3.2" resolve to the same HAProxy 3.2.z — the pin is a no-op but preserves intent for future upgrades. Upgrading from 5.0 with `"2.8"` is blocked. |
| 5.2 | unset<br/>`"3.4"`<br/>`"3.2"` | 3.4, 3.2 | OCP 5.2 introduces HAProxy 3.4. Unset/"3.4" use HAProxy 3.4. "3.2" uses the latest HAProxy 3.2.z built for this release. Unset and "3.4" use the same version; "3.4" pins during upgrades. |
```

Author: Updated the version matrix almost as is, adding a few more context about pins from the version selection rules section.

**Key Points**:

> - Each OCP release ships with exactly 2 HAProxy container images

Reviewer: Is this still accurate? 5.1 seem to only support 1 in your previous example

Author: We're moving now to "at most 2 versions per OCP release" approach now. Updated.

> - Each OCP release ships with exactly 2 HAProxy container images

Reviewer 3: Related - you could add this principle to help illustrate

```suggestion
- Each OCP release ships with at most 2 HAProxy container images
- A release has no previous HAProxy version when it neither introduces a new minor nor carries one forward from a prior release.
```

but only add if you think it adds value.

Author: Sure, added it as well. Thanks.

- Unset field always uses the latest/default HAProxy for that release
- **Version transition pattern**: When an OpenShift release introduces a new
  HAProxy minor version, the previous pinnable version is dropped in the next
  OpenShift release. For example, OCP 5.0 introduces HAProxy 3.2, making "3.2"
  a pinnable version in subsequent releases. Support for "2.8" (HAProxy 2.8,
  the previous pinnable version) is limited to OCP 5.0 only and is dropped in
  OCP 5.1.
- The OpenShift releases introducing a new HAProxy minor version (e.g., 5.0)
  allow pinning to the current version for upgrade stability, in addition to

>   pinning to a previous supported release (e.g., 2.8).

Reviewer 3: I think this "Version Transition Pattern" might be causing confusion. This is describing the N-1 version. We should switch back to ELC now that we have cleared the terminology/releases as public:

```suggestion
- **Version transition pattern**: New HAProxy minor versions are introduced
  on ELC releases. The previous pinnable version is carried until the next
  ELC release, which introduces a new minor and drops the oldest. For
  example, OCP 5.2 (ELC) introduces HAProxy 3.4, and "3.2" remains
  pinnable on 5.3 and 5.4. When OCP 5.5 (ELC) introduces HAProxy 3.6,
  "3.2" is dropped and "3.4" becomes the pinnable version.
  Note: HAProxy 2.8 is a special case — support is limited to OCP 5.0
  only and dropped in 5.1 due to its upstream end-of-life status.
- The OpenShift releases introducing a new HAProxy minor version (e.g., 5.2)
  allow pinning to the current version for upgrade stability, in addition to
  pinning to a previous supported release (e.g., "3.2").
```

    alt Version not available
        Operator->>IC: Set degraded condition
        Operator->>Admin: Report error in status
    else Version available
        Operator->>Router: Update deployment with version
        Router->>Router: Restart with new HAProxy binary
        Router->>Operator: Report readiness
        Operator->>IC: Update status with active version
        IC->>Admin: Display updated version in status
    end
```

### API Extensions

This enhancement modifies the existing IngressController CRD
(`ingresscontrollers.operator.openshift.io/v1`) to add a new optional field for specifying
the HAProxy version.

The proposed API type definitions for OCP 4.22:

```go
type IngressControllerSpec struct {
	// ... existing fields ...

	// haproxyVersion specifies the HAProxy version to use for this
	// IngressController.
	//
	// This field is available in OpenShift 4.22 as an API-only backport with no
	// operator implementation. Setting this field on OpenShift 4.22 allows
	// administrators to pin HAProxy 2.8 before upgrading to OpenShift 5.0, where
	// the operator will honor this setting.
	//
	// Valid values for OpenShift 4.22:

> 	// - Unset (default): Uses HAProxy 2.8 (the default for OpenShift 4.22)

Reviewer: The explicit godoc is great - we would need to remember to update the validation every release, but hopefully this would help us note which versions we support per reelase

Author: We made some minor changes to godoc after reviewing some parts of the EP. Wondering if you can have another look on it.

Reviewer: The documentation reads fine to me - I guess my main concern here is that assuming this is the plan, like we'd have to go something like:

1. introduce the FG as techpreview in 5.0
2. implement the feature in 5.0, add tests for feature promotion
3. graduate the feature in 5.0
4. backport the graduated feature and API to 4.22

Which will be quite an involved process, given that we have ~1month for the 5.0 timeline, this will be quite tight.

appropriate HAProxy binaries and dependencies.

#### Standalone Clusters

This enhancement is fully applicable to standalone clusters and represents
the primary use case. Administrators can manage HAProxy versions
independently across multiple IngressControllers.

#### Single-node Deployments or MicroShift

For Single-Node OpenShift (SNO) deployments, this enhancement provides the
same benefits as standalone clusters. Resource consumption is minimal as
only the selected HAProxy binary and its dependencies are loaded.

For MicroShift, this enhancement may not be directly applicable as
MicroShift has a different ingress architecture. If MicroShift adopts
IngressController resources in the future, this enhancement could be
extended to support it.

#### OpenShift Kubernetes Engine

This enhancement works with OpenShift Kubernetes Engine (OKE) as it relies
on standard IngressController resources which are available in OKE.

### Implementation Details

#### Chosen Implementation: External HAProxy Images with Sidecar Deployment

This implementation deploys HAProxy as a separate sidecar container alongside the
main router container. Each supported HAProxy version is packaged in its own

> dedicated container image. Each OCP release includes 2 distinct HAProxy sidecar

Reviewer 3: ```suggestion
dedicated container image. Each OCP release includes at most 2 distinct HAProxy sidecar
```

   HAProxy 3.4 remains pinnable until the next release that introduces a new
   HAProxy minor version.
   
   **Version progression for OCP 5.0-5.2**:
   - OCP 5.0: Ships HAProxy 3.2 LTS (introduces HAProxy 3.2)
   - OCP 5.1: Ships HAProxy 3.2.z (z-stream patches only)
   - OCP 5.2: Ships HAProxy 3.4 LTS (introduces HAProxy 3.4)
   
   **Pinned version behavior**: When administrators pin to a HAProxy version
   (e.g., `haproxyVersion: "3.2"` on OCP 5.1 or 5.2), they receive HAProxy 3.2.z
   from the OpenShift release that originally introduced that HAProxy minor
   version. The pinned HAProxy version continues to receive z-stream patches as
   they are backported to that release stream. For example, pinning to "3.2" on
   OCP 5.1 or 5.2 uses HAProxy 3.2.z from the OCP 5.0 release stream (which
   introduced HAProxy 3.2), and continues to be updated with security and bug
   fixes.
   
   This strategy ensures that:
   - HAProxy version changes are controlled and predictable
   - Releases between major HAProxy bumps provide stability with security and
     bug fixes
   - Pinned versions remain secure through continued z-stream patching

>    - No EOL upstream HAProxy versions are supported long-term

Reviewer 3: Per our discussion (feel free to alter):

```suggestion
   **Pinned version behavior**: When administrators pin to a HAProxy minor version
   (e.g., `haproxyVersion: "3.2"` on OCP 5.1 or 5.2), they receive the latest
   HAProxy 3.2.z available for that OpenShift release. The pinned value is a
   y-stream contract — the z-stream may change between OCP releases as patches
   are applied. For example, pinning to "3.2" on OCP 5.0 might use HAProxy
   3.2.1, while pinning to "3.2" on OCP 5.1 might use HAProxy 3.2.2.
   
   This strategy ensures that:
   - HAProxy minor version changes are controlled and predictable
   - No frozen release streams or separate CVE backport streams are required
   - The API remains open for explicit z-stream pinning (e.g., "3.2.1") in
     the future if needed
```

Author: Looks good, applied as is. Thanks.

> 

Reviewer 3: Something that is forward looking, but important to not lock ourselves out of this design:

```suggestion

   **Future Version opt-in**: In addition to the pinnable previous LTS described
   above, we may ship a newer HAProxy version on any release —
   including non-ELC releases — as an opt-in option for customers to
   validate ahead of the next ELC. The default (unset) continues to use
   the default HAProxy version for that release, while the newer version
   is available for early adoption by explicitly setting it. In the future,
   the system may be extended to support more than 2 versions simultaneously,
   such as a previous LTS, the current standard, and an upcoming version
   available for early validation.
```

## Graduation Criteria

We will introduce the feature as TechPreviewNoUpgrade during the OpenShift 5.0
development, with the goal of graduating it to GA before it is released.

### Dev Preview -> Tech Preview

N/A.

### Tech Preview -> GA

- Payload tests pass with HAProxy selection enabled
- All known regressions are fixed
- Extensive testing including upgrade scenarios
- End-to-end tests included in standard conformance suites
- Support procedures documented for troubleshooting version issues
- Limitations, such as migration criterias, are documented

### Removing a deprecated feature

N/A - This is a new feature.

## Upgrade / Downgrade Strategy

**Upgrades**:

When upgrading an OpenShift cluster:
1. IngressControllers with `haproxyVersion` unset will automatically use the
   new HAProxy version from the upgraded OpenShift release, except in the case
   of OpenShift 4.22 migration (see the special case below).
2. IngressControllers with `haproxyVersion: "X.Y"` will preserve the
   specified HAProxy version, provided it remains a supported value for the
   target release.
3. If a pinned version would become unsupported in the target release, the

>    cluster-ingress-operator will set the ingress ClusterOperator's `Upgradeable`

Reviewer: This is probably better as a preflight check (when that gets implemented, which I think is still the plan) eventually. Maybe we can make a note of that here.

Author: I changed from the status condition to the preflight check, let me know if I understood you correctly.

## Proposal

This enhancement proposes adding a new field to the IngressController API
that allows administrators to specify which HAProxy version to use. When
the field is unset (the default), it always uses the default HAProxy version
from the current OpenShift release.

When an administrator specifies an OpenShift release version, the
IngressController will use the HAProxy version that shipped with that
specific OpenShift release. The cluster-ingress-operator will manage the
deployment of the appropriate HAProxy binary and its dependencies (pcre,
openssl, FIPS libraries) to support the selected version.

At most 2 distinct HAProxy versions are supported simultaneously: a pinned
version from a previous supported release and a new version if introduced in
the current OpenShift release. The feature begins with OCP 5.0.

**Version Selection Rules**:

Each OpenShift release supports at most 2 distinct HAProxy versions. The
supported values for `haproxyVersion` are:

- **OCP 4.22**: Supports unset and "2.8"
  - Unset and "2.8" uses HAProxy 2.8
  - Note: API-only backport - field available but no operator implementation;

>     setting "2.8" allows pinning HAProxy 2.8 when upgrading to 5.0

Reviewer 3: ```suggestion
    setting "2.8" allows pinning HAProxy 2.8 when upgrading to 4.23 or 5.0
```

> 

Reviewer 3: I think 4.23 is quite important too.

```suggestion

- **OCP 4.23**: Supports unset, "2.8" and "3.2"
  - Unset uses HAProxy 3.2
  - "3.2" uses HAProxy 3.2 (explicitly pins version during cluster upgrades)
  - "2.8" uses HAProxy 2.8 from the 4.22 release
  - Note: EUS-to-ELC users should upgrade to 4.23 to have the ability to select 
    between HAProxy 2.8 and 3.2 before proceeding to 5.1.
```

    pins the version during cluster upgrades

- **OCP 5.3**: Supports unset and "3.4"
  - Unset and "3.4" uses the latest HAProxy 3.4.z available for this OpenShift
    release (pins the minor version; z-stream floats with each OCP release)
  - Note: "3.2" is dropped; HAProxy 3.2 support ends; a future HAProxy version
    may be added for customers to validate

- **OCP 5.4**: Supports unset and "3.4"
  - Unset and "3.4" uses the latest HAProxy 3.4.z available for this OpenShift
    release (pins the minor version; z-stream floats with each OCP release)
  - Note: A future HAProxy version may be added for customers to validate

- **OCP 5.5**: Supports unset, "3.4" and "3.6"
  - Unset uses HAProxy 3.6
  - "3.6" uses HAProxy 3.6 (explicitly pins version during cluster upgrades)
  - "3.4" uses the latest HAProxy 3.4.z available for this OpenShift release
  - Note: Unset and "3.6" use the same HAProxy image; specifying "3.6" explicitly
    pins the version during cluster upgrades

**Supported Version Matrix**:

The following table shows which `haproxyVersion` field values are valid for
each OpenShift release, and which HAProxy images are shipped with each release:

> | OCP Release | Valid Field Values | HAProxy Versions Available | Notes |

Reviewer 3: *nit* the bulleted list and the table are essentially the same thing, but the table stops early at 5.2. Not a big deal. I'd probably just get rid of one of them. I like the table, but looks like the bulleted list has more versions, I'd just drop the table.

Maybe you could use a table like this just to help visualize (i'm a very visual person):

```suggestion
| OCP  | 2.8 | 3.2 | 3.4 | 3.6 |
|------|-----|-----|-----|-----|
| 4.22 | D   |     |     |     |
| 4.23 | P   | D   |     |     |
| 5.0  | P   | D   |     |     |
| 5.1  |     | D   | ?   |     |
| 5.2  |     | P   | D   |     |
| 5.3  |     |     | D   | ?   |
| 5.4  |     |     | D   | ?   |
| 5.5  |     |     | P   | D   |

D = default (unset), P = pinnable, ? = may be added for early validation
```

Author: Much better, this is how the table should look like since the very beginning I think. Note that I added a center alignment and an extra "/ HAProxy".

| 5.1 | unset<br/>`"3.2"` | 3.2 | HAProxy 2.8 dropped. Unset and "3.2" resolve to the latest HAProxy 3.2.z built, but "3.2" pin preserves intent for future upgrades. Upgrading from 5.0 with "2.8" is blocked. |
| 5.2 | unset<br/>`"3.4"`<br/>`"3.2"` | 3.4, 3.2 | OCP 5.2 introduces HAProxy 3.4. Unset and "3.4" resolve to the same HAProxy 3.4.z, but "3.4" pin preserves intent for future upgrades. "3.2" uses the latest HAProxy 3.2.z built for this release. |

**Key Points**:
- Each OCP release ships with at most 2 HAProxy container images
- A release has no previous HAProxy version when it neither introduces a new
  minor nor carries one forward from a prior release
- Unset field always uses the default HAProxy for that release
- **Version transition pattern**: New HAProxy minor versions are introduced
  on ELC releases. The previous pinnable version is carried until the next ELC
  release, which introduces a new minor and drops the oldest. For example, OCP
  5.2 (ELC) introduces HAProxy 3.4, and "3.2" remains pinnable on 5.3 and 5.4.
  When OCP 5.5 (ELC) introduces HAProxy 3.6, "3.2" is dropped and "3.4"

>   becomes the pinnable version.

Reviewer 3: So the bullet points up top describe one-release-drop:

```
  One-release-drop-after-ELC (bullet points):
  | OCP  | 3.2 | 3.4 | 3.6 |
  |------|-----|-----|-----|
  | 5.0  | D   |     |     |
  | 5.1  | D   |     |     |
  | 5.2  | P   | D   |     |
  | 5.3  |     | D   |     |
  | 5.4  |     | D   |     |
  | 5.5  |     | P   | D   |
```

But this here says Carry-until-next-ELC:

```
Carry-until-next-ELC (Key Points):
  | OCP  | 3.2 | 3.4 | 3.6 |
  |------|-----|-----|-----|
  | 5.0  | D   |     |     |
  | 5.1  | D   |     |     |
  | 5.2  | P   | D   |     |
  | 5.3  | P   | D   |     |
  | 5.4  | P   | D   |     |
  | 5.5  |     | P   | D   |
```

Both models provide an identical experience for ELC-to-ELC hoppers — the version pinned at the current ELC is always valid at the next ELC regardless of which carry model we use. I prefer the one-release-drop model described in the bullet points for a couple of reasons:

- **Encourages timely migration**: Dropping the previous version sooner avoids letting release-by-release upgraders defer a transition they'll need to make anyway.
- **Frees up the second version slot on non-ELC releases**: Releases like 5.3 and 5.4 would only carry one version, leaving room for early validation — we could introduce an upcoming HAProxy version ahead of the next ELC, letting users test it before it becomes the default.

Author: In fact I didn't realize the phrasing was against some other parts of the EP, including the bullet points and the version matrix. Adjusted now.

     container performs route watching, configuration generation, and HAProxy
     management through the admin socket, which are less resource-intensive
     operations compared to traffic handling.
   - **Total**: 125m CPU and 320Mi memory requests across both containers,
     which represents an increase of 25%.
   
   This allocation strategy prioritizes HAProxy performance for traffic handling
   while providing sufficient resources for the router's control plane
   operations.

9. **HAProxy Version Bump Strategy**: HAProxy minor version bumps follow a
   controlled strategy to provide predictable upgrade paths and maintain version
   stability:
   
   **When and why we bump**: OpenShift adopts new HAProxy LTS (Long Term Support)
   versions when they become available from upstream and when the currently
   supported HAProxy version approaches end-of-life. HAProxy LTS releases occur
   approximately once per year. OpenShift releases that do not introduce a new
   default HAProxy version preserves the option to pin the same HAProxy version
   introduced on a previous release, receiving only z-stream patch updates (bug

>    fixes and security patches) from their respective release streams.

Reviewer 3: ```suggestion
   default HAProxy version preserves the option to pin the same HAProxy version
   introduced on a previous release. The pinned version receives the latest
   patches available for that minor version in each OpenShift release.
```

builds for each affected version. Clearly document supported versions and
deprecation timelines. Provide warnings in the IngressController status when
using older versions. Consider implementing alerts when versions reach
end-of-support.

**Risk 3**: Dependency conflicts between HAProxy versions and their required
libraries (pcre, openssl, FIPS), including potential incompatibilities when
running older libraries from previous OCP releases on newer kernels.

**Mitigation**: Thoroughly test each supported version with its dependencies
on each supported kernel version. Each HAProxy sidecar image is self-contained
with all required dependencies, eliminating library isolation concerns.
Implement robust validation during version selection. Document tested
library/kernel combinations and known incompatibilities.

**Risk 4**: Complexity in troubleshooting when different IngressControllers
run different HAProxy versions.

**Mitigation**: Clearly expose the HAProxy version in IngressController
status. Add metrics and logging to identify which version is running.
Include version information in the must-gather artifacts.

> 

Reviewer 3: Thoughts on documenting this risk for y-stream floating?

```suggestion

**Risk 5**: Y-stream floating means the HAProxy z-stream version can change
between OCP releases without the administrator explicitly opting in. For
example, an administrator who sets `haproxyVersion: "3.2"` on OCP 5.0 might
run HAProxy 3.2.1, then after upgrading to OCP 5.1, the same `"3.2"` setting
could resolve to HAProxy 3.2.3. While z-stream updates are generally
backwards-compatible, a behavioral change in a z-stream bump could
unexpectedly affect a user's configuration.

**Mitigation**: Z-stream updates for HAProxy LTS releases are limited to bug
fixes and security patches. The API remains open for explicit z-stream pinning
(e.g., `haproxyVersion: "3.2.1"`) in the future if this proves to be a
concern in practice.
```

Author: Good addition, added risk 5 suggestion as is.

  clusters
- Providing formal support for custom HAProxy versions or images not shipped
  with supported OpenShift releases (though the mechanism may technically
  allow custom versions in an unsupported configuration - see Open Question #11)
- Allowing selection of specific HAProxy version numbers (e.g., "2.8.5") in
  the standard supported configuration
- Providing version selection for other ingress components beyond HAProxy
  itself (router code, templates, etc.)
- Supporting more than 2 distinct HAProxy versions simultaneously
- Supporting OCP versions older than the previous supported release

## Proposal

This enhancement proposes adding a new field to the IngressController API
that allows administrators to specify which HAProxy version to use. When
the field is unset (the default), it always uses the default HAProxy version
from the current OpenShift release.

When an administrator specifies an OpenShift release version, the
IngressController will use the HAProxy version that shipped with that
specific OpenShift release. The cluster-ingress-operator will manage the
deployment of the appropriate HAProxy binary and its dependencies (pcre,

> openssl, FIPS libraries) to support the selected version.

Reviewer 3: *nit* I think part wasn't updated from the OCP version model.

```suggestion
When an administrator specifies a HAProxy minor version (e.g., "3.2"), the
IngressController pins that HAProxy version, preserving it across future
cluster upgrades even if the default version changes. When the field is
unset, the IngressController uses the default HAProxy version for the
current OpenShift release. The cluster-ingress-operator manages the
deployment of the appropriate HAProxy binary and its dependencies (pcre,
openssl, FIPS libraries) to support the selected version.
```

Author: Good catch, fixed now.

	//   upgrades to future OpenShift releases
	// - "2.8": Uses HAProxy 2.8 from OpenShift 4.22 (migration support, will
	//   be dropped in the next OpenShift release)
	//
	// If a specific HAProxy version is set and would become unsupported in a
	// target cluster upgrade, a preflight check would block the cluster upgrade,
	// blocking the cluster upgrade until this field is updated to unset or a
	// supported version.
	//
	// +optional
	// +openshift:enable:FeatureGate=IngressControllerMultipleHAProxyVersions
	HAProxyVersion HAProxyVersion `json:"haproxyVersion,omitempty"`
}

type IngressControllerStatus struct {
	// ... existing fields ...

	// effectiveHAProxyVersion reports the HAProxy version currently in use by
	// this IngressController. This reflects the resolved value of the
	// spec.haproxyVersion field. When omitted, the effective value has not yet
	// been resolved by the operator or the feature is not enabled for this cluster.
	//
	// Examples for OpenShift 5.0:
	// - "3.2": Using HAProxy 3.2
	// - "2.8": Using HAProxy 2.8
	//
	// +optional
	// +openshift:enable:FeatureGate=IngressControllerMultipleHAProxyVersions

> 	EffectiveHAProxyVersion HAProxyVersion `json:"effectiveHAProxyVersion,omitempty"`

Reviewer 3: Personally, I think that this would provide more value as status to the user as a full x.y.z resolution of the version. E.g. `3.2` is selected, but this resolves to `3.2.1` in 5.1.

However, that involves a process for the CIO to resolve the full HAProxy version, which I can't think of a easy query to to do that. Router probably needs to expose an endpoint to return the effective HAProxy value. 

Just a thought and definitely extra credit at this point, it's probably too complicated to implement. Changing the type from `HAProxyVersion` to `String` in the future should be a compatible change, so it's something we could always do later.

Author: This would be a good addition. I though about it but also about what should be changed to make it in a proper way, so better not to count with it for now.

We could also hardcode the version in the operator since we have full control on what we are deploying, but it would be a fragile approach and needs maintenance on every release.

We could also change the type now, so it is flexible and already accept the new approach in case we find the time to implement, test and revise it. Otoh we'd need to change the API godoc as well, so, in case it fits, we can change EP and API later.

changes incrementally across multiple OpenShift releases.

As a platform operations team, I want to monitor and manage HAProxy versions
across multiple IngressControllers at scale, so that I can ensure
consistency and track version adoption across the cluster.

### Goals

- Enable cluster administrators to select HAProxy versions for individual
  IngressControllers independently of OpenShift cluster version
- Support preservation of HAProxy versions from the current OpenShift
  release and up to 2 previous releases to facilitate LTS-to-LTS migrations
- Allow testing new HAProxy versions on dedicated IngressControllers before
  production deployment
- Maintain compatibility with dynamic HAProxy compilation and required
  dependencies (pcre, openssl, FIPS libraries)
- Default to latest available HAProxy version when no explicit version is
  selected

### Non-Goals

- Backporting HAProxy versions from newer OpenShift releases to older
  clusters
- Allowing selection of arbitrary HAProxy versions not shipped with
  supported OpenShift releases
- Providing version selection for other ingress components beyond HAProxy

>   itself (router code, templates, etc.)

Reviewer 5: Are we saying that together with HAProxy we will select the corresponding router and config template?

Because otherwise there could be some issue if we change only the HAProxy version.

Author: In fact having router+haproxy is also an approach, I just updated the EP with all the proposals, this one being the 3rd option.

IngressController will use the HAProxy version that shipped with that
specific OpenShift release. The ingress-controller-operator will manage the
deployment of the appropriate HAProxy binary and its dependencies (pcre,
openssl, FIPS libraries) to support the selected version.

Up to 3 distinct HAProxy versions will be supported simultaneously,
allowing administrators to migrate from one LTS OpenShift version to the
next while preserving HAProxy versions during the transition period.

### Workflow Description

**cluster administrator** is a human user responsible for managing
OpenShift cluster infrastructure and upgrades.

#### Selecting a HAProxy Version

1. The cluster administrator reviews the current OpenShift release notes and
   identifies the HAProxy version shipped with the new release.
2. The cluster administrator creates or updates an IngressController resource,
   specifying the desired HAProxy version in the new API field (e.g.,
   `haproxyVersion: "OCP-4.21"` or `haproxyVersion: "Current"`).
3. The ingress-controller-operator validates the requested version is
   available and supported.
4. The operator updates the IngressController deployment to use the

>    specified HAProxy version and its matching dependencies.

Reviewer 5: I suggest to specify in a better way how we would like to manage this point.

The Cluster Ingress Operator today selects a single router image, not a HAProxy version, so there will be different options:

- one router image containing multiple HAProxy builds,
- multiple router images selected per IngressController,
- or a sidecar/init-container/artifact injection model 

The EP should describe with a higher level of detail this point

Author: Added 3 distinct proposals handling exactly the three options you suggested here.

    Operator->>Operator: Validate version availability
    alt Version not available
        Operator->>IC: Set degraded condition
        Operator->>Admin: Report error in status
    else Version available
        Operator->>Router: Update deployment with version
        Router->>Router: Restart with new HAProxy binary
        Router->>Operator: Report readiness
        Operator->>IC: Update status with active version
        IC->>Admin: Display updated version in status
    end
```

### API Extensions

This enhancement modifies the existing IngressController CRD
(`operator.openshift.io/v1`) to add a new optional field for specifying
the HAProxy version.

The proposed API field:

```go
// HAProxyVersion specifies which HAProxy version to use for this
// IngressController. Valid values are:
// - "Current" (default): Use the latest HAProxy version available in this
//   OpenShift release
// - "OCP-X.Y": Use the HAProxy version from OpenShift release X.Y
//
// Supports the current OpenShift release and up to 2 prior releases,
// starting from "OCP-4.23".
//
// +optional
// +kubebuilder:default="Current"
// +openshift:enable:FeatureGate=SelectableHAProxyVersion

> HAProxyVersion string `json:"haproxyVersion,omitempty"`

Reviewer 5: I think that a free-form string with comments saying valid values are "Current" and "OCP-X.Y" could be not enough. I'll add at least:

- `Pattern` validation for the accepted forms,
- a corresponding status field that reports effective and requested values separately.

Author: Cool, just added both of them.

selection applies to IngressControllers in both the management cluster and
guest clusters. The ingress-controller-operator in each context manages the
appropriate HAProxy binaries and dependencies.

#### Standalone Clusters

This enhancement is fully applicable to standalone clusters and represents
the primary use case. Administrators can manage HAProxy versions
independently across multiple IngressControllers.

#### Single-node Deployments or MicroShift

For Single-Node OpenShift (SNO) deployments, this enhancement provides the
same benefits as standalone clusters. Resource consumption is minimal as
only the selected HAProxy binary and its dependencies are loaded.

For MicroShift, this enhancement may not be directly applicable as
MicroShift has a different ingress architecture. If MicroShift adopts
IngressController resources in the future, this enhancement could be
extended to support it.

#### OpenShift Kubernetes Engine

This enhancement works with OpenShift Kubernetes Engine (OKE) as it relies
on standard IngressController resources which are available in OKE.

### Implementation Details/Notes/Constraints

> The implementation requires the following high-level code changes:

Author: Missing high-level comment about the approaches to package HAProxy and isolate its dynamic dependencies. Wondering if it fits here to comment all the possibilities, choosing one of them later when starting to implement.

Author: Added 3 proposals.

     references
   - Report the active HAProxy version in the IngressController status

4. **HAProxy Binary Management**: Since HAProxy is compiled dynamically, the
   implementation must:
   - Package multiple HAProxy versions with their dependencies (pcre,
     openssl, FIPS libraries) in the router image, in a dedicated HAProxy
     image, or as separate artifacts
   - Ensure the correct dynamic libraries are available for each version
     if installed in the same image
   - Maintain compatibility matrices for HAProxy versions and their
     dependencies

5. **Version Validation**: Implement validation to ensure:
   - Only supported OpenShift release versions can be specified
   - Maximum of 3 distinct versions are maintained
   - Requested versions are available in the current cluster

6. **Upgrade/Downgrade Handling**: Implement logic to handle cluster
   upgrades and downgrades according to the specified version policy:
   - "Current" always uses the latest available version
   - Specific versions are preserved if available in the target release
   - Fall back to default if a version becomes unsupported when
     Downgrading

>    - Deny cluster upgrades when any IngressController references a

Reviewer 5: Related to this point I've got the following questions:

- Which component sets Upgradeable=False?
- What exact reason/message?
- How is the target release computed?
- What is the skew behavior during a partially rolled out update?
- Why is downgrade fallback acceptable but upgrade denial required?

Author: I'm not very skilled on upgrade validations, so wondering if you or someone skilled on it can share some thoughts on how it works, so I can convert this into how it applies to the haproxy version validation.

Reviewer 3: @davidesalerno

1. CIO sets Upgradeable=False on it's IngressController that has the "old" version, and then CIO propagates that up to the `ClusterOperator` object (the `ClusterOperator` upgradegradable=false is what blocks the upgrade)
2. We can decide later - something like "Current HAProxy version is not available in next x.y OCP version"
3. Target release should just mean the next OCP version (N+1). EP could be updated to be more direct. We don't allow skipping versions: even EUS-to-EUS upgrades step through N+1. So the upgrade validation only needs to check against the immediate next release.
4. HAProxy version skew during rollout isn't a big concern. Each pod handles traffic independently with no shared state between them. If some pods are running 4.22's HAProxy alongside 5.0's HAProxy during the transition, both serve requests correctly from the same route config.
   - As for skew between the router controller and HAProxy: they're in the same pod in the sidecar model, and Kubernetes updates pods atomically. You can't update one container without the other.
6. We don't have to support downgrades

I don't think this level of detail needs to be in the EP - but I hope this helps.

multiple HAProxy versions to router pods (data plane). The operator version
determines which HAProxy versions are available, not the router pod version.

**Multi-Version Support Window**:
The operator maintains compatibility with HAProxy versions from the current
OpenShift release and up to 2 previous releases. During a rolling upgrade,
different router pods may temporarily run different HAProxy versions, which
is acceptable as each IngressController operates independently.

**API Compatibility**:
The new `haproxyVersion` field is optional and gated behind a feature gate.
Older versions of the operator (before the feature) will ignore this field.
Newer versions of the operator will handle both the presence and absence of
the field gracefully.

**Kubelet Compatibility**:
This enhancement does not involve kubelet changes. HAProxy version selection
is entirely managed by the ingress-controller-operator and router pods.

## Operational Aspects of API Extensions

### SLIs for API Extensions

This enhancement modifies an existing CRD (IngressController) but does not
add webhooks or aggregated API servers. The existing SLIs for

> IngressController resources apply:

Author: Using field validation and effective version in status to handle this.
